Proposal: Variadics (#2240)
carbon-language/carbon-lang: docs/design/variadics.md

segment-wise. For example, suppose we are trying to find the type of `z` in this
code:

```carbon
fn F[... each T:! type]((... each x: Optional(each T)), (... each y: i32)) {
  let z: auto = (0 as f32, ... each x, ... each y);
}
```

We proceed by finding the type of each segment. The type of `0 as f32` is `f32`,
by the usual non-variadic typing rules. The type of `... each x` is
`... Optional(each T)`, because `Optional(each T)` is the declared type of
`each x`, and the type of a pack expansion is a pack expansion of the type of
its body.

The type of `... each y` is more complicated. Conceptually, it consists of some
number of repetitions of `i32`. We don't know exactly how many repetitions,
because it's implicitly specified by the caller: it's the arity of the second
argument tuple. Effectively, that arity acts as a hidden deduced parameter of
`F`.

So to represent this type, we need two new pseudo-syntaxes:

-   `‖each X‖` refers to the deduced arity of the pack expansion that contains
    the declaration of `each X`.
-   `⟪E; N⟫` evaluates to `N` repetitions of `E`. This is called a _arity
    coercion_, because it coerces the expression `E` to have arity `N`. `E` must

>     not contain any pack expansions, each-names, or pack literals.

Reviewer: Note: use of term "pack literal" before it is defined.

Author: Good catch. I've marked it explicitly as a forward reference, because I don't see a way to avoid it.

Reviewer: I don't see that change

Author: Sorry, that reply wasn't supposed to be published yet. I'm not sure how that happened.

Combining the two, the type of `... each y` is `... ⟪i32; ‖each y‖⟫`. Thus, the
type of `z` is `(f32, ... Optional(each T), ... ⟪i32; ‖each y‖⟫)`.

Now, consider a modified version of that example:

```carbon
fn F[... each T:! type]((... each x: Optional(each T)), (... each y: i32)) {

>   let (each z: auto) = (0 as f32, ... each x, ... each y);

Reviewer: Does this need a `...` on the left?
```suggestion
  let (... each z: auto) = (0 as f32, ... each x, ... each y);
```

}
```

`each z` is a pack, but it has the same elements as the tuple `z` in our earlier
example, so we represent its type in the same way, as a sequence of segments:
`⟬f32, Optional(each T), ⟪i32; ‖each y‖⟫⟭`. The `⟬⟭` delimiters make this a
_pack literal_ rather than a tuple literal. Notice one subtle difference: the
segments of a pack literal do not contain `...`. In effect, every segment of a
pack literal acts as a separate loop body. As with the tuple literal syntax, the

> pattern literal pseudo-syntax can also be used in patterns.

Reviewer: Is there a "pattern literal pseudo-syntax" or should this be:
```suggestion
pack literal pseudo-syntax can also be used in patterns.
```

-   A statement pack expansion requires no further typechecking, because
    statements don't have types.
-   An `...and` or `...or` expression has type `bool`, and every segment of the
    operand's type pack must have a type that's convertible to `bool`.
-   For a `...` tuple element expression or pattern, the segments of the
    operand's type pack become segments of the type of the enclosing tuple.

> **TODO:** Discuss typechecking `...expand`.

### Typechecking patterns

A _full pattern_ consists of an optional deduced parameter list, a pattern, and
an optional return type expression.

A pack expansion pattern has _fixed arity_ if it contains at least one usage of
an each-name that is not a parameter of the enclosing full pattern. Otherwise it
has _deduced arity_. A tuple pattern can have at most one segment with deduced
arity. For example:

```carbon
class C(... each T:! type)
  fn F[... each U:! type](... each t: each T, ... each u: each U);
}
```

In the signature of `F`, `... each t: each T` has fixed arity, and
`... each u: each U` has deduced arity, because the arity of `each U` is
determined by the arguments passed to `F`, but the arity of `each T` is

> determined by the arguments passed to `C`, before the call to `F`.

Reviewer: This is a pretty involved sentence. How about we break it up?

```suggestion
In the signature of `F`, `... each t: each T` has fixed arity, since the arity is determined by the arguments passed to `C`, before the call to `F`. On the other hand, `... each u: each U` has deduced arity, because the arity of `each U` is determined by the arguments passed to `F`.
```

must match at least one element.)

When the pattern is heterogeneous, the merging process may be more complex. For
example:

```carbon
fn ZipAtLeastOne[First:! type, ... each Next:! type]
    (first: Vector(First), ... each next: Vector(each Next))
    -> Vector((First, ... each Next));
```

During typechecking, we transform that function signature to the following form:

```carbon
fn ZipAtLeastOne[... ⟬First, each Next⟭:! ⟪type; ‖each next‖+1⟫]
    (... each __args: Vector(⟬First, each Next⟭))
    -> Vector((... ⟬First, each Next⟭));
```

We can then rewrite that by replacing the pack of names `⟬First, each Next⟭`
with an invented name `each __Args`, so that the function has only one
parameter:

```carbon
fn ZipAtLeastOne[... each __Args:! ⟪type; ‖each next‖+1⟫]
    (... each __args: Vector(each __Args))
    -> Vector((... each __Args));
```

We can replace a name pack with an invented each-name only if all of the
following conditions hold:

-   The name pack doesn't use any name more than once. For example, we can't
    apply this rewrite to `⟬X, each Y, X⟭`.
-   The name pack contains at least one each-name. For example, we can't apply

>     this rewrite to `⟬X, Y⟭`.

Reviewer: Is there a situation where you can perform a name pack replacement with more than one each-name (or more than one with deduced arity)?

Author: I don't think so. Changed "at least one" to "exactly one".

non-variadic typechecking.

> **TODO:** Extend this approach to fall back to a complementary approach, where
> the pattern and scrutinee trade roles: we maximally merge the scrutinee tuple,
> while requiring each segment to have a single scalar component, and then
> merge/split the pattern tuple to match it, without requiring pattern tuple
> segments to have a single scalar component. This isn't quite symmetric with
> the current approach, because when processing the scrutinee we can't merge
> deduced parameters (scrutinees don't have any), but we can invent new `let`
> bindings.

## Appendix: Type system formalism

A _pack literal_ is a comma-separated sequence of segments, enclosed in `⟬⟭`
delimiters. A pack literal can appear in an expression, pattern, or name
context, and every segment must be valid in the context where the pack literal
appears (for example, the segments of a pack literal in a name context must all
be names). Pack literals cannot be nested, and cannot appear outside a pack
expansion.

### Explicit deduced arities

In this formalism, deduced arities are explicit rather than implicit, so Carbon
code must be desugared into this formalism as follows:

> For each pack expansion pattern, we introduce a binding pattern `__N:! Arity` as

Reviewer: Is this always done, or only if needed? It seems like it might not be needed, for example when it has fixed arity (and so there is an earlier arity parameter that governs the arity of this pack expansion pattern).

(see your example at the end of this doc)

Author: Good point. Fixed.

associative.

### Typing and shaping rules

The shape of an AST node within a pack expansion is determined as follows:

-   The shape of an arity coercion is the value of the expression after the `;`.
-   The shape of a pack literal is the concatenation of the arities of its
    segments.
-   The shape of an each-name expression is the shape of the binding pattern
    that declared the name.
-   If a binding pattern's name and type components have the same number of
    segments, and each name segment is an each-name if and only if the
    corresponding type segment's shape is not 1, then the shape of the binding
    pattern is the shape of the type expression. Otherwise, the binding pattern
    is ill-shaped.
-   For any other AST node:
    -   If all the node's children have shape 1, its shape is 1.
    -   If there is some shape `S` such that all of the node's children have
        shape either 1 or `S`, its shape is `S`.
    -   Otherwise, the node is ill-shaped.

> **TODO:** The "well-shaped" rules as stated are slightly too restrictive. For
> example, `⟬each X, Y⟭: ⟪Z; N+1⟫` is well-shaped, and `(⟬each X, Y⟭, ⟪Z; N+1⟫)`

> > is well-shaped if the shape of `each Y` is `N`.

Reviewer: ```suggestion
> is well-shaped if the shape of `each X` is `N`.
```

_Pack expansion splitting:_ If `E` is a segment and `S` is a sequence of
segments, `...⟬E, S⟭` reduces to `...E, ...⟬S⟭`.

_Pack expanding:_ If `F` is a function, `X` is an utterance that does not
contain pack literals or each-names, and `⟬P1, P2⟭` and `⟬Q1, Q2⟭` both have the
shape `(S1, S2)`, then `F(⟬P1, P2⟭, X, ⟬Q1, Q2⟭, ⟪Y; S1+S2⟫)` reduces to
`⟬F(P1, X, Q1, ⟪Y; S1⟫), F(P2, X, Q2, ⟪Y; S2⟫)⟭`. This rule generalizes in
several dimensions:

-   `F` can have any number of non-pack-literal arguments, and any positive
    number of pack literal arguments, and they can be in any order.
-   The pack literal arguments can have any number of segments (but the
    well-shapedness requirement means they must have the same number of
    segments).
-   `F()` can be any expression syntax other than `...`, not just a function
    call. For example, this rule implies that `⟬X1, X2⟭ * ⟬Y1, Y2⟭` reduces to
    `⟬X1 * Y1, X2 * Y2⟭`, where the `*` operator plays the role of `F`.
-   `F()` can also a be a pattern syntax. For example, this rule implies that
    `(⟬x1: X1, x2: X2⟭, ⟬y1: Y1, y2: Y2⟭)` reduces to
    `⟬(x1: X1, y1: Y1), (x2: X2, y2: Y2)⟭`, where the tuple syntax `( , )` plays

>     the role of `F`.

Reviewer: Perhaps split pattern and tuple syntax into separate bullet points?

Author: I don't understand; there's no additional fact here about tuples that would call for a separate bullet point. I just had to choose some specific pattern syntax for the example, so I chose tuple pattern syntax because it seems like the closest parallel to the earlier examples. I've tweaked the phrasing to make that a little more explicit. If that's not enough, can you explain what problem you're trying to solve?

(I've added a bullet point below about binding patterns because I realized it was missing. That's not intended to address your comment, but maybe it will help?)

_Coercion expanding:_ If `F` is a function, `S` is a shape, and `Y` is an
expression that does not contain pack literals or arity coercions,
`F(⟪X; S⟫, Y, ⟪Z; S⟫)` reduces to `⟪F(X, Y, Z); S⟫`. As with pack expanding,
this rule generalizes:

-   `F` can have any number of non-arity-coercion arguments, and any positive
    number of arity coercion arguments, and they can be in any order.
-   `F()` can be any expression syntax other than `...` or pack literal
    formation, not just a function call. Unlike pack expanding, coercion
    expanding does not apply if `F` is a pattern syntax.

_Coercion removal:_ `⟪E; 1⟫` reduces to `E`.

_Tuple indexing:_ Let `I` be an integer template constant, let `X` be a tuple
segment, and let `Ys` be a sequence of tuple segments.

-   If the arity `A` of `X` is less than `I+1`, then `(X, Ys).(I)` reduces to
    `(Ys).(I-A)`.
-   Otherwise:
    -   If `X` is not a pack expansion, then `(X, Ys).(I)` reduces to `X`.
    -   If `X` is of the form `...⟬⟪E; S⟫⟭`, then `(X, Ys).(I)` reduces to `E`.

### Other equivalences

Unless otherwise specified, all expressions in these rules must be free of side
effects.

_Coercion merging:_

> -   `⟪E; M⟫, ⟪E; N⟫` is equivalent to `⟪E; M, N⟫`.

Reviewer: Shouldn't you only have `+` to the right of a `;` in a `⟪E; ...⟫`? Having both `+` and `,` means there are a lot more rules than would otherwise be needed. And a lot of the rules are just what was already stated about arities.

Author: Yeah, this used to be necessary because arity coercions used to be "shape coercions", but I don't think it's needed now. Fixed.

and there's an additional premise that this union is consistent. If there are no
"equals" or "convertible to" relations in the premises, the binding map in the
conclusion is empty. When that pattern applies, we will usually leave the
binding maps implicit (as in the original formulation of the above rule). One
major exception is the following rule:

_Binding introduction:_ Let `X` be a deduced parameter with type `T`. If the
type of `E` is convertible to `T` given binding map `M`, and `M` ∪ {(`X`, `E`)}
is consistent, then `X` equals `E` given a binding map `M` ∪ {(`X`, `E`)}.

The deduction rules that are specific to variadic types are as follows:

-   _Expansion convertibility:_ `...T` is convertible to `...U` if the arity of
    `U` equals the arity of `T`, and all scalar components of `T` are
    convertible to all scalar components of `U`.
-   _Shape equality:_ Let `(S1s)`, `(S2s)`, `(S3s)`, and `(S4s)` be shapes.
    `(S1s, S2s)` equals `(S3s, S4s)` if `(S1s)` equals `(S3s)` and `(S2s)`
    equals `(S4s)`.

#### Example

Suppose we are typechecking the following function call:

```carbon
fn F[T:! type](x: Vector(T));
F(() as Vector(i32));

> ```

Reviewer: I think there has been some scope creep here: you are defining deduction in general, independent of variadics.

Author: OK, I've removed the non-variadic-specific parts.

(To be clear, "scope creep" wasn't how this happened; earlier versions of the design made more invasive changes to the convertibility and deduction rules, so they couldn't be explained as a couple of additional rules on top of an unstated baseline)

We need to verify that `Vector(i32)` is convertible to `Vector(T)`. We can
construct a proof by applying the above inference rules, as follows:

1. (By identity) `type` equals `type` given a binding map {}.
2. (By equality implies convertibility) `type` is convertible to `type` given a
   binding map {}.
3. (By binding introduction) `T` equals `i32` given a binding map {(`T`,
   `i32`)}.
4. (By parameterized type equality) `Vector(T)` equals `Vector(i32)` given a
   binding map {(`T`, `i32`)}.
5. (By equality implies convertibility) `Vector(T)` is convertible to
   `Vector(i32)` given a binding map {(`T`, `i32`)}.

Notice that as a byproduct, this proof gives us a binding map with the deduced
values of the parameters.

Conceptually, the typechecker constructs this proof by searching backwards from
desired conclusion without considering binding maps, and then running the
resulting proof forward to compute the final binding map.

### Deduction algorithm

A full pattern is in _normal form_ if it contains no pack literals, and every

> arity coercion is fully expanded. Note that all user-written full patterns are

Reviewer: You use the term "expansion" a lot above. Could you explain what you mean here with an example? Something like: "and every arity coercion is fully expanded (and so looks like ___ not ___)"

Author: How's this?

in normal form. Note also that by construction, this means that the type of the
body of every pack expansion has a single scalar component. The _canonical form_

> of a full pattern is the unique normal form (if any) that is "maximally merged",

Reviewer: Naively "maximally merged" sounds like the opposite of "arity coercion is fully expanded"

Author: I've added an example to try to clarify that.

> **TODO:** Specify the algorithm more precisely. In particular, discuss how to
> rewrite `A` as needed to make the shapes line up, but don't rewrite `F` after
> canonicalization.

Typechecking for pattern match operations other than function calls is defined
in terms of typechecking a function call: We check a scrutinee type `S` against
a pattern `P` by checking `__F(S,)` against a hypothetical function signature
`fn __F(P,)->();`.

> **Future work:** Extend this approach to support merging the argument list as
> well as the parameter list.

#### Canonicalization algorithm

The canonical form can be found by starting with a normal form, and
incrementally merging an adjacent singular parameter type into the variadic
parameter type.

For example, consider the following function:

```carbon
fn F[First:! type, Second:! type, ... each Next:! type]
    (first: Vector(First), second: Vector(Second),
     ... each next: Vector(each Next)) -> (First, Second, ... each Next);
```

First, we desugar the implicit arity:

```carbon
fn F[__N:! Arity, First:! type, Second:! type, ... each Next:! ⟪type; __N⟫]
    (first: Vector(First), second: Vector(Second),

>      ... each next: Vector(each Next)) -> (First, Second, ... each Next);

Reviewer: (continued from before)
In this example, there are two pack expansion patterns (`... each Next` and `... each next`), but only one implicit arity.

names. This kind of sequence is called a _pack_. Packs are very similar to tuple
values in many ways, but they are not first-class values -- in particular, packs
do not have types, and no expression evaluates to a pack. The _arity_ of a pack
is a compile-time value representing the number of values in the sequence. All
arguments of a given expansion must have the same arity (which we will also
refer to as the arity of the expansion).

A pack expansion cannot occur within another pack expansion.

A pack expansion can be thought of as a kind of loop that executes at compile
time (specifically, monomorphization time), where the expansion body is
implicitly parameterized by an integer value called the _pack index_, which
ranges from 0 to one less than the arity of the expansion. The pack index is
implicitly used as an index into the expansion argument packs. This is easiest
to see with statement pack expansions. For example, if `a`, `x`, and `y` are
tuples of size 3, then `... expand a += expand x * expand y;` is roughly
equivalent to

```carbon
for (let template i:! i32 in (0, 1, 2)) {
  let product: auto = x[i] * y[i];

>   a[i] += product;

Reviewer: ```suggestion
  a[i] += x[i] * y[i];
```

}
```

Notice, however, that this notional rewritten form is not valid Carbon code,

> because the expressions `x[i]`, `y[i]`, and `z[i]` may have different types

Reviewer: ```suggestion
because the expressions `a[i]`, `x[i]`, and `y[i]` may have different types
```

A pack expansion must contain one or more _expansion arguments_, which are
marked by `expand` or `each`. `expand` is a prefix unary expression operator
with the same precedence as `*`. The operand of `each` is always an identifier
name or a binding pattern, so it does not have a precedence. For example, in the
loop condition `...and expand iters != each vector.End()` in the body of `Zip`,
`expand iters` and `each vector` are expansion arguments. Note that `each` can
also appear in a deduced parameter list, as discussed below.

The _arity_ of an expansion argument is a compile-time value representing the
number of elements it evaluates to (or the number of values it matches, if it's
a pattern). Every pack expansion must contain at least one expansion argument,
and all arguments of a given expansion must have the same arity (which we will
also refer to as the arity of the expansion). If any expansion argument is a
non-expression pattern, the entire pack expansion is a pattern. In particular,
this means that the expansion arguments of a `...{` expansion must be
expressions.

Pack expansions can be nested only if the inner expansion is within one of the

> outer expansion's arguments. For example,

Reviewer 2: It's not clear to me what this means -- what are the arguments of the outer expansion?

Is the idea here that you can't nest a pack expansion in another pack expansion "at the top level"? Eg, `(..., ..., expand each tuple)` wouldn't be permitted, but `(..., (..., expand each tuple))` would be. Or that you can't form a pack-of-packs, so `expand each tuple` would itself be invalid?

Reviewer 2: Oh, perhaps this means "within the outer expansion's <ins>expansion</ins> arguments"? Which I think would translate into a "no packs of packs" restriction.

Author: Yeah, it was intended to be "no packs of packs", but implemented as a syntactic restriction. However, I'm realizing the rest of the proposal has shifted out from under this rule: this exception can only apply to `expand`, because a pack expansion in the type of an `each` binding would lead to a pack-of-packs situation (and the attendant ambiguities about which expansion the arguments belong to). Furthermore, it doesn't apply to `expand` patterns because we don't support `expand` in patterns anymore -- which means the motivating use case given here doesn't work. Consequently, I've switched to a total ban on nesting.

_block-contents_ N times, with `$I` ranging from 0 to N - 1.

An expression of the form "`...and` _operand_" is evaluated as follows: a
notional `bool` variable `$R` is initialized to `true`, and then "`$R = $R and`
_operand_" is executed up to N times, with `$I` ranging from 0 to N - 1. If at
any point `$R` becomes false, this iteration is terminated early. The final
value of `$R` is the value of the expression.

An expression of the form "`...or` _operand_" is evaluated the same way, but
with `or` in place of `and`, and `true` and `false` transposed.

An expression of the form "`...,` _operand_" evaluates to a sequence of N
values, where the k'th value is the value of _operand_ where `$I` is equal to
k - 1.

An expression of the form "`expand` _argument_" evaluates to "_argument_
`[$I]`".

An expression of the form "`each` _identifier_", where _identifier_ names a
variadic binding, evaluates to the `$I`th value that it was bound to (indexed
from zero).

## Typechecking expressions and statements

### Generalized tuple types

The `...,` operator lets us form tuples out of sequences whose size is not known
during typechecking. For example, in this code:

```carbon

> fn F[each T:! type](x: (..., each i32), ..., each y: Optional(each T)) {

Reviewer 2: `each i32` doesn't seem to be valid per the earlier rules, as `i32` is not a variadic binding nor a binding pattern. Should this be:

```suggestion
fn F[each T:! type](x: (..., i32), ..., each y: Optional(each T)) {
```

... where the pack expansion in the type of `x` doesn't contain any expansion arguments? While that would normally be problematic, I think it's OK in the type of a pattern because we can deduce the arity based on usage.

Author: Yes, `each i32` is definitely an error, but I've fixed it in a different way.

On a technical level it seems straightforward to support patterns like `x: (..., i32)`, but the readability angle needs more careful consideration. So far we've been sticking to the principle that deduced types are always marked by a variable, placeholder, or some locally-recognizable token sequence (like `;]` in a deduced-size array type like `[i32;]`). Your suggestion would introduce a case where deduction occurs by default, unless it is suppressed by an `each` or `expand` somewhere within the type expression (so you need to read the whole thing to verify that deduction is taking place).

In some ways this is the tradeoff of moving away from arrays as the starting point for variadics: it gives us a simpler syntax by removing the position where the pack size would go, which leaves us with no natural syntactic lever for deducing the pack size without deducing the element types (this will also be an issue if we want to support homogeneous variadic members).

Reviewer: Isn't the only thing being deduced is the arity of the tuple, and that fact is being marked by `...`?

Author: `...` is also used in a lot of non-pattern contexts where deduction doesn't happen, unlike `auto` and `;]`, which dilutes its value as a marker of deduction. Even in patterns it's not a fully reliable signal -- e.g. there's no arity deduction in a pattern like `x: (... expand (i32, i32))` (which is silly but valid).

illustration, the notation `<V, N>` represents a pack component with
representative `V` and arity `N`, `$I` represents the pack index, and given a
variadic binding `B`, `|B|` represents the arity of `B`, and `B/$I` represents
the `$I`th value bound by `B`.

So, continuing the earlier example, the type of `z` is represented symbolically
as `(<i32, SizeOf(x)>, <f32, 1>, <Optional(T/$I), |T|>)`.

A pack component is _variadic_ if its arity is unknown, and _singular_ if its
arity is known to be 1 and its representative does not refer to the pack index.
In contexts where all pack components are known to be singular, we will
sometimes refer to them as "elements". Pack values are always assumed to be
normalized, meaning that every component is either singular or variadic. This is
always possible because if a component's arity is known to be some fixed value N
other than 1, we can replace it with N singular components. The _shape_ of a
tuple type is the sequence of arities of its components, so the shape of the
type of `z` is `(SizeOf(x), 1, |T|)`.

In order to index into a tuple, all of the components of its type must be

> singular, so that we can determine the type of the indexing expression.

Reviewer 2: We could be a bit more permissive: in order to index into a tuple with index `I`, the shape of the tuple must begin with at least `I + 1` 1s, so that the `I`th element is known to be singular.

after another. As a result, a pack type is never the type of any run-time value.

## Pattern semantics

`,...` expansions can also appear in patterns. The semantics are chosen to
follow the general principle that pattern matching is the inverse of expression
evaluation, so for example if the pattern `(..., expand x: auto)` matches some
scrutinee value `s`, the expression `(..., expand x)` should be equal to `s`.
These semantics are implemented at monomorphization time, so all types are known
constants, and all tuple elements are singular.

A tuple pattern can contain no more than one subpattern of the form "`...,`
_operand_". When such a subpattern is present, the N elements of the pattern
before the `...,` expansion are matched with the first N elements of the
scrutinee, and the M elements of the pattern after the `...,` expansion are
matched with the last M elements of the scrutinee. If the scrutinee does not
have at least N + M elements, the pattern does not match.

The remaining elements of the scrutinee are iteratively matched against
_operand_, in order. In each iteration, `$I` is equal to the index of the

> scrutinee element being matched, minus N.

Reviewer 2: The way this is phrased sounds like we first match the elements before the `...,`, then the ones after it, then the ones in the middle. I find that a bit surprising -- if there are T total elements, I'd expect we would first match the first N elements in order, then the T - N - M elements in the middle in order, then the last M elements in order.

// a tuple of the i'th elements of the input vectors.
fn Zip[... each ElementType:! type]
      (... each vector: Vector(each ElementType))
      -> Vector((... each ElementType)) {
  ... var each iter: auto = each vector.Begin();
  var result: Vector((... each ElementType));
  while (...and each iter != each vector.End()) {
    result.push_back((... each iter));
    ... each iter++;
  }
  return result;
}
```

A _pack_ is a sequence of a fixed number of values called "elements", which may
be of different types. Packs are very similar to tuple values in many ways, but
they are not first-class values -- in particular, packs do not have types, and
no expression evaluates to a pack. The _arity_ of a pack is a compile-time value
representing the number of values in the sequence.

A pattern of the form "`...` _subpattern_" is called a _pack expansion pattern_.
It can only appear as part of a tuple pattern (or an implicit parameter list),
and it matches a sequence of tuple elements if each element matches
_subpattern_. Since _subpattern_ will be matched against multiple scrutinees (or
none) in a single pattern-matching operation, it cannot contain a ordinary

> binding patterns. However, it can contain _pack binding patterns_, which are

Reviewer 2: ```suggestion
none) in a single pattern-matching operation, it cannot contain any ordinary
binding patterns. However, it can contain _pack binding patterns_, which are
```
or
```suggestion
none) in a single pattern-matching operation, it cannot contain ordinary
binding patterns. However, it can contain _pack binding patterns_, which are
```
or
```suggestion
none) in a single pattern-matching operation, it cannot contain an ordinary
binding pattern. However, it can contain _pack binding patterns_, which are
```

Usages of the `each` keyword (in other words, pack binding patterns and usages
of pack bindings) are called _expansion sites_. The _arity_ of an expansion site
is a symbolic value representing the arity of the underlying pack. Expansion
sites can only occur in the body of a _pack expansion_, which is an instance of
one of the following syntactic forms:

-   A statement of the form "`...` _statement_".
-   A tuple expression element of the form "`...` _expression_", with the same
    precedence as `,`.
-   A tuple pattern element of the form "`...` _pattern_", with the same
    precedence as `,`.
-   An implicit parameter list element of the form "`...` _pattern_", with the
    same precedence as `,`.
-   An expression of the form "`...` `and` _expression_" or "`...` `or`
    _expression_", with the same precedence as `and` and `or`.

The `...` token can also occur in a tuple expression element of the form "`...`
`expand` _expression_", with the same precedence as `,`. However, that syntax is
not considered a pack expansion, and has its own semantics: _expression_ must
have a tuple type, and "`...` `expand` _expression_" evaluates _expression_ and

> treats its elements as elements of the enclosing tuple literal. This is

Reviewer 2: I think this isn't necessarily quite the same thing as saying that this is restricted syntax that forms and immediately expands a pack. In particular:
```
fn F[...template each Tuple:! type](...each tuple: Tuple) {
  ... G(...expand each tuple);
}
```
... would mean `F((a, b), (c, d, e))` succeeds and calls `G(a, b)` then `G(c, d, e)`, whereas if `...expand` formed and then expanded a pack, we'd reject due to the nested pack expansion.

This isn't an objection to treating `...expand` as not being a pack thing, just an observation that there's something slightly unusual happening here, that may go a little beyond what we'd allow if it *were* a pack thing.

If we're OK with that example, I think we should be clearer whether `G(... ...expand each tuple)` is allowed, as well as things like `H(...expand ...expand multi_level_tuple)`. The description of `...expand` as having a particular precedence suggests this might be OK, but the above rule about these things appearing as tuple elements suggests that it might not be.

Author: Good point. I'd prefer say that it's not a pack expansion because that makes it easier to explain the behavior of pack expansions (e.g. expansion sites always begin with `each`). But I think we should disallow those forms of nesting anyway, in order to preserve our ability to generalize `expand` later, so I've added that restriction.

especially useful for using non-literal tuple values as function call arguments:

```carbon
fn F(x: i32, y: String);
fn MakeArgs() -> (i32, String);

F(...expand MakeArgs());
```

`...and`, `...or`, and `...expand` can be trivially distinguished with one token

> of lookahead, and the other can be distinguished from each other by the context

Reviewer 2: ```suggestion
of lookahead, and the other meanings of `...` can be distinguished from each other by the context
```

they appear in. As a corollary, if the nearest enclosing delimiters around a
`...` are parentheses, they will be interpreted as forming a tuple rather than
as grouping. Thus, expressions like `(... each ElementType)` in the above
example are tuple literals, even though they don't contain commas.

By convention, `...` is always followed by whitespace, except that `...and`,
`...or`, and `...expand` are written with no whitespace between the two tokens.
This serves to emphasize that the keyword is not part of the expansion body, but
rather a modifier on the syntax and semantics of `...`.

In a pack expansion, the statement, expression, or pattern embedded in the
expansion is called the _body_ of the expansion, and a pack expansion body must
contain at least one expansion site. All sites of a given expansion must have
the same arity (which we will also refer to as the arity of the expansion).

> A pack expansion cannot occur within another pack expansion.

Reviewer 2: What's the motivation for this? I don't think nested pack expansions have proven problematic in C++, but then again C++ doesn't do full type-checking prior to pack expansion, and doesn't have any "packs out of nowhere" syntax like `expand tuple`. (In contrast, packs-of-packs might present some implementation difficulties, and `expand each blah` would result in a pack-of-packs.)

Author: See the "Allow nested pack expansions" section in p2240.md.

Also, nested pack expansions may be unproblematic as of C++20, but they seem to have been problematic for some pending C++ proposals. For example, the most recent (unpublished) draft of the expansion statements proposal removes support for expanding over packs, because it would lead to ambiguity in cases like this example (credited to one Richard Smith):

```c++
template<typename ...T>
void f(T ...v) {
  g([&](auto y) {
    template for (auto x : v) { /*...*/ } // Pack expansion or not?
  }(v)...);
}
```

Here an expansion statement is nested within an expansion expression, and it's unclear whether the "expansion site" `v` on the commented line belongs to the outer or inner expansion.

A pack binding cannot be used in the same pack expansion that declares it. In

> almost all cases, a binding that violates this rule can be changed to a non-pack

Reviewer 2: Do you have a specific case in mind for this "almost"? I've been trying to find one and haven't succeeded.

Author: I don't have anything specific in mind, I just haven't fully convinced myself that there are no such cases. I've rephrased a bit.

binding, because pack bindings are only necessary when you need to transfer a
pack from one pack expansion to another.

A pack expansion can be thought of as a kind of loop that executes at compile
time (specifically, monomorphization time), where the expansion body is
implicitly parameterized by an integer value called the _pack index_, which
ranges from 0 to one less than the arity of the expansion. The pack index is
implicitly used as an index into the expansion site packs. This is easiest to
see with statement pack expansions. For example, if `a`, `x`, and `y` are pack
bindings with arity 3, then `... each a += each x * each y;` is roughly
equivalent to

```carbon
for (let template i:! i32 in (0, 1, 2)) {
  a[:i:] += x[:i:] * y[:i:];
}
```

Here we are using `[::]` as a hypothetical pack indexing operator for purposes
of illustration; packs cannot actually be indexed in Carbon code. Note also that
this rewritten form would not typecheck under the usual rules, because the
expressions `a[:i:]`, `x[:i:]`, and `y[:i:]` may have different types depending

> on the value of `i`.

Reviewer 2: I think it might actually be OK that these expressions have different types depending on `i`, because `i` is a template binding.

Author: Possibly, although we haven't specified that yet AFAIK. In any event, the `template` is really a relic from when `a`, `x`, and `y` were tuples, and it can't be a template binding in the general case anyway, because the range it iterates over won't necessarily be a template constant. I've dropped it.

```carbon
// Toy example of mixing variadic and non-variadic parameters.
// Takes an i64, any number of f64s, and then another i64.
fn MiddleVariadic(first: i64, ... each middle: f64, last: i64);
```

```carbon
// Toy example of using the result of variadic type deduction.
fn TupleConcat[... each T1: type, ... each T2: type](
    t1: (... each T1), t2: (... each T2)) -> (... each T1, ... each T2) {
  return (...expand t1, ...expand t2);
}
```

## Semantics

### Expressions and statements

In all of the following, N is the arity of the pack expansion being discussed,
and `$I` is a notional variable representing the pack index. These semantics are
implemented at monomorphization time, so the value of N is a known integer
constant. Although the value of `$I` can vary during execution, it is
nevertheless treated as a constant.

A statement of the form "`...` _statement_" is evaluated by executing
_statement_ N times, with `$I` ranging from 0 to N - 1.

An expression of the form "`...and` _expression_" is evaluated as follows: a
notional `bool` variable `$R` is initialized to `true`, and then "`$R = $R and`
_expression_" is executed up to N times, with `$I` ranging from 0 to N - 1. If

> at any point `$R` becomes false, this iteration is terminated early. The final

Reviewer 2: Is the compile-time iteration terminated early -- that is, we don't substitute higher values of `$I` into the expression and type-check them -- or does the compile-time iteration always complete but the run-time execution may terminate early?

Author: At this point I'm only describing the execution semantics (I've renamed the section to make that more explicit). The specification in "Iterative typechecking" below currently implies that compile time iteration is never terminated early, but I hadn't really considered the alternative.

Now that I have, though, I tentatively think that's the right result: it seems like it would be really surprising if an `...and` expression could typecheck when the manually-expanded form would not (I assume we'll always typecheck both sides of an ordinary `and` expression). That does mean we'll probably need to fall back on recursive template instantiation to implement the Carbon equivalents of `std::conjunction` and `std::disjunction`, but my hope is that those will be less necessary in Carbon, and it seems possible that the way they're used won't lead to the factor-of-N performance problems we get with recursive variadics in some cases.

Possibly if we support `...&` (in the interface-combining sense of `&`), it would make sense for that to short-circuit at typechecking time?

> value of `$R` is the value of the expression.

Reviewer 2: The description of the loop here, while precise, might be slightly obscuring what we're doing and why. I think it might be useful to say that this is equivalent to `true and E0 and E1 and ... and EN-1`, where `Ei` is _expression_ with `$I` having the value `i`. (Maybe even start with something like that and then go on to say, more precisely, that we perform this compile-time loop.)

Author: In this section I'm trying to precisely restate the intuitive descriptions in the previous section, but I've made changes to the previous section along the lines you suggest. How's that?

An expression of the form "`...or` _expression_" is evaluated the same way, but
with `or` in place of `and`, and `true` and `false` transposed.

A tuple expression element of the form "`...` _expression_" evaluates to a
sequence of N values, where the k'th value is the value of _operand_ where `$I`
is equal to k - 1.

An expression of the form "`each` _identifier_", where _identifier_ names a pack
binding, evaluates to the `$I`th value that it was bound to (indexed from zero).

### Pattern matching

`...` expansions can also appear in patterns. The semantics are chosen to follow
the general principle that pattern matching is the inverse of expression
evaluation, so for example if the pattern `(... each x: auto)` matches some
scrutinee value `s`, the expression `(... each x)` should be equal to `s`. These
semantics are implemented at monomorphization time, so all types are known

> constants, and all tuple elements are singular.

Reviewer 2: ```suggestion
constants, and all tuple elements are [singular](#generalized-tuple-types).
```
We've not defined "singular" yet, and this is the first mention of the term.

Author: I've rephrased to avoid the forward reference.

`$I` represents the pack index, and given a pack binding `B`, `|B|` represents
the arity of `B`, and `B[:$I:]` represents the `$I`th value bound by `B`.

So, continuing the earlier example, the type of `z` is represented symbolically
as `(<i32, |x|>, <f32, 1>, <Optional(T[:$I:]), |T|>)`.

A segment is _variadic_ if its arity is unknown, and _singular_ if its arity is
known to be 1 and its representative does not refer to the pack index. In
contexts where all segments are known to be singular, we will sometimes refer to
them as "elements". Segments are always assumed to be normalized, meaning that
every segment is either singular or variadic. This is always possible because if
a segment's arity is known to be some fixed value N other than 1, we can replace
it with N singular segments. The _shape_ of a tuple type is the sequence of
arities of its segments, so the shape of the type of `z` is `(|x|, 1, |T|)`.

In order to index into a tuple with subscript `I`, the tuple type's segment
sequence must start with at least `I` singular segments, so that we can
determine the type of the indexing expression. Note that this rule applies only

> to user-written subscript operations, not to the notional `[:$I:]` operations

Reviewer 2: ```suggestion
to user-written subscript operations with an index that is not template-dependent, not to the notional `[:$I:]` operations
```
I would expect you *can* index with a `template i:! i32`, because we don't do the indexing until we monomorphize.

Author: I would argue that this rule also applies if the index is template-dependent: in that case, the indexing expression is likewise template-dependent, so it is not typechecked until monomorphization, at which point all arities are known, and so all segments are normalized to be singular, trivially satisfying this requirement.

it has a _type pack_ that represents the sequence of types it takes on over the
course of the iteration. Just as a pack is not quite a value, a type pack is not
quite a type, but type packs relate to packs in the same way that types relate
to values.

An expansion site's pack is composed of a sequence of elements from a tuple, so
its type pack consists of the types of those elements. However, as discussed
above, at typechecking time we don't necessarily know the type of each tuple
element, or even how many elements there are -- we only know the sequence of
segments. As a result, a type pack is represented as a sequence of segments.

In some cases, the typechecker may need to reason about the values contained in
a pack, as well as their types. For example:

```
fn f[...each A:! I1, ...each B:! I2]((... each a: each A), (...each b: each B)) {
  let (... each X:! type) = (...each A, ...each B);
  let (... each x: each X) = (...each a, ...each b);
}
```

In order to represent the declared type of `x`, the typechecker needs to model
the fact that `X` consists of the concatenation of `A` and `B`, even though it
does not yet know the arity or contents of either pack. It does this by

> representing packs as sequences of segments. For example, `Y` can be represented

Reviewer 2: What's `Y`? Should this be `X`?

as `(<A[:$I:], |A|>, <B[:$I:], |B|>)`.

Similarly, the typechecker sometimes needs to reason symbolically about values
with generalized tuple types (like the value of `(...each A, ...each B)`).
These, too, are represented as sequences of segments.

Since type packs are sequences of segments, typechecking must iterate over those
segments' representatives rather than over the (unknown) individual element
types. To ensure that this is valid, we require all sites of a given expansion
to have the same shape. We determine the shapes of the expansion sites as
follows:

-   If the expansion contains any non-deducing usages of pack bindings, the
    bindings they name must all have the same shape, and all other expansion
    sites are deduced to have the same shape.
-   Otherwise, we invent a local symbolic variable `N`, and all expansion sites
    are treated as having the shape `(N,)` (representing a single segment with

>     unknown arity).

Reviewer 2: I think we should say that it's an error if `N` cannot then be deduced. Eg:
```carbon
// should be an error, can't deduce arity of `x`
var (...each x: i32);
```
... should be rejected. The exception would be when we're forming a pattern that will later have its arity determined by some scrutinee:
```carbon
// OK, arity deduced from call
fn F(...each x: i32);
// OK, arity deduced from initializer
var (...each x: i32) = (1, 2, 3);
// OK, arity deduced from scrutinee
match (1, 2, 3) {
  case (...each x: i32) => {}
}
```
... or from another pack, by this rule:
```carbon
fn F(...each T:! type) {
  // OK, arity of `x` is deduced as `|T|` from later expansion site.
  var (...(each x: i32, each y: each T));
}
```
Maybe that rule doesn't belong here specifically, but it feels like we need to say that somewhere :)

Author: Good point. How's this?

The type packs of expressions and patterns in the expansion body are then
determined by iterative typechecking: within the k'th typechecking iteration,
typechecking and symbolic evaluation behave as they do during non-variadic
typechecking, except that:

-   For every expression/pattern in the expansion, the type of the k'th segment
    of its type pack takes the place of its type (on both reads and writes).
    // - We reuse (rather than attempt to recompute) any type information that

>     was // computed in the course of the shape-checking described above.

Reviewer 2: Looks like a bullet got `// commented out` here and then prettier'd?

```carbon
fn F(a: i32, ... each b: i32, c: i32);

fn G(... each x: i32) {
  F(1, 2, ... each x);
}
```

If `G` is called with no arguments, the `2` will match with `c`, and otherwise
the `2` will match with an element of `b`. Similarly, if `G` is called with one
or more arguments, the last argument will match `c`, and the remaining elements
(if any) will match values from `b`. However, at type-checking time we don't
know how many arguments the caller will pass, so we don't know which will occur.
On the other hand, the `1` will always match `a`.

In general, we want type checking to fail if any possible monomorphization of
the generic code would fail to typecheck. In this case that means we want type
checking to fail if any of the potential argument-parameter mappings could fail
to typecheck after monomorphization. Furthermore, for reasons of readability as
well as efficiency, we want type checking to fail if any two potential mappings
would deduce inconsistent values for any deduced parameter. However, in general
this is intractable, because in the worst case the number of distinct ways to
map symbolic arguments to parameters is ${2n \choose n}$ for n variadic

> arguments, which is only a factor of $\sqrt{n}$ away from exponential.

Reviewer 2: ```suggestion
arguments, which is exponential.
```
a^n / sqrt(n) = O(b^n), for any 1 < b < a.

Author: Oh, cool! In that case I think I'd also drop the discussion of combinatorics. This is moot due to the overhaul of this whole section, but see commit 14b97b362 for how it will look if the bigger overhaul is reverted.

Introducing type deduction further complicates the situation. For example:

```carbon
fn H[each T:! type](a: i32, ... each b: each T, c: String) -> (... each T);

external impl P as ImplicitAs(i32);
external impl Q as ImplicitAs(String);

> fn I(each x: i32, each y: f32, each z: String) {

Reviewer 2: ```suggestion
fn I((... each x: i32), (... each y: f32), (... each z: String)) {
```

  var result: auto = H(... each x, {} as P, ... each y, {} as Q, ... each z);
}
```

Here, the deduced type of `result` can have one of four different forms. The

> most general case is `(<i32, |x|>, <P, 1>, <f32, |y|>, <Q, 1>, <String, |z|>)`,

Reviewer 2: ```suggestion
most general case is `(<i32, |x|-1>, <P, 1>, <f32, |y|>, <Q, 1>, <String, |z|-1>)`,
```

and the other three cases are formed by omitting the first two and/or last two
segments (corresponding to the cases where `x` and/or `z` do not match any
arguments). Extending the type system to support deduction that splits into
multiple cases would add a fearsome amount of complexity to the type system.

> #### Identifying potential matchings

Reviewer 2: This seems like a lot of complication, and I wonder whether it's justified. I'd be inclined to reject any case where we can't map each argument to exactly one parameter based on shape alone -- so singular parameters must have singular arguments, fixed-arity pack expansions must have matching fixed-arity arguments, and at most one variable-arity pack parameter can remain, which gets the same shape as the rest of the arguments.

Are there important use cases that don't work in that kind of approach?

Reviewer: I'm assuming you would allow multiple arguments to map to the single variadic parameter? This is needed to call a variadic with a non-variadic argument list.

One of the motivating use cases for the current approach was to support the case where functions requiring at least one argument, but the caller puts the extra argument at the end while the callee puts the extra parameter at the beginning (or the other way around).

Author: As discussed on Discord and elsewhere, `Min(... each x, y)` is an example of something that wouldn't work with that approach, but your question led me to realize that the rules I had didn't support that case either. However, your approach can be made to work if we extend the type system and syntax to allow `Min` to be written with a single variadic parameter that's required to have at least one element, so that's roughly what I've done here (commit b16a5d9cc7 if you want to see it in isolation from the other comment responses). I'll follow up with changes to p2240.md to discuss the tradeoffs of this approach.

Author: I've now updated p2240.md, so this should now be fully ready for review.

}
```

A _pack_ is a sequence of a fixed number of values called "elements", which may
be of different types. Packs are very similar to tuple values in many ways, but
they are not first-class values -- in particular, packs do not have types, and
no expression evaluates to a pack. The _arity_ of a pack is a compile-time value
representing the number of values in the sequence.

A pattern of the form "`...` _subpattern_" is called a _pack expansion pattern_.
It can only appear as part of a tuple pattern (or an implicit parameter list),
and it matches a sequence of tuple elements if each element matches
_subpattern_. Since _subpattern_ will be matched against multiple scrutinees (or
none) in a single pattern-matching operation, it cannot contain ordinary binding
patterns. However, it can contain _pack binding patterns_, which are binding
patterns that begin with `each`, such as `each ElementType:! type`. A pack
binding pattern can match an arbitrarily large number of times, and binds to a
pack consisting of all the matched values. A usage of a pack binding always has
the form "`each` _identifier_".

By default, a pack binding pattern can match any number of times, but `each` can

> optionally be followed by "`(>=` _integer-literal_ `)`", which constrains the

Reviewer 3: While I was reading your nice proposal, I was thinking about `(>=` lakes the left operand. Is it possible to put `>=` next to `...`?

```cpp
...>=1 each param
```

In this way, it doesn't require parentheses, and the left operand of `>=` is `...`. Thanks.

Author: Yes, that's possible, and arguably it would be more accurate, because it actually does constrain the whole expansion. The main reason I attached it to `each` is that in that position, there's no ambiguity: `each` can only be followed by an identifier or an arity constraint, which are easy to distinguish. With `...` it's more subtle, because `...` can be followed by many things, including some operators. It can't otherwise be followed by `>=`, so it's not formally ambiguous, but it seems like it could easily lead to confusion for human readers, who might expect `...>=` to be related to `>=` in the same way that `...and` is related to `and`. Also, I'm already kind of uncomfortable with how we've given `...` so many syntactic variants (like `...and`, `...or`, and `...expand`), so I'm particularly reluctant to create yet another.

If we did attach it to the `...`, I think I'd want to keep the parentheses, in order to emphasize that this is nothing like `...and` and that the `1` is not part of the expansion body.

## Table of contents

-   [Basics](#basics)
    -   [Additional examples](#additional-examples)
-   [Execution Semantics](#execution-semantics)
    -   [Expressions and statements](#expressions-and-statements)
    -   [Pattern matching](#pattern-matching)
-   [Typechecking](#typechecking)
    -   [Tuples, packs, segments, and shapes](#tuples-packs-segments-and-shapes)
    -   [Iterative typechecking of pack expansions](#iterative-typechecking-of-pack-expansions)
    -   [Typechecking patterns](#typechecking-patterns)
    -   [Typechecking pattern matches](#typechecking-pattern-matches)
-   [Appendix: Type system formalism](#appendix-type-system-formalism)
    -   [Explicit deduced arities](#explicit-deduced-arities)
    -   [Typing and shaping rules](#typing-and-shaping-rules)
    -   [Reduction rules](#reduction-rules)
    -   [Equivalence, equality, and convertibility](#equivalence-equality-and-convertibility)
    -   [Pattern match typechecking algorithm](#pattern-match-typechecking-algorithm)
        -   [Canonicalization algorithm](#canonicalization-algorithm)
-   [Alternatives considered](#alternatives-considered)
-   [References](#references)

<!-- tocstop -->

> ## Basics

Reviewer 4: I think it would be useful to have a few sub-sections of this section to help the reader keep track of where things are... At the least, the point where this switches from pack expansion expressions / statements into pack _patterns_ is a bit easy to miss I find.

Author: OK, how's this?

Reviewer 4: Nice!

A pack expansion or `...expand` expression cannot contain another pack expansion
or `...expand` expression.

An each-name cannot be used in the same pack expansion that declares it. In most
if not all cases, an each-name that violates this rule can be changed to an
ordinary name, because each-names are only necessary when you need to transfer a
pack from one pack expansion to another.

A pack expansion expression or statement can be thought of as a kind of loop
that executes at compile time (specifically, monomorphization time), where the
expansion body is implicitly parameterized by an integer value called the _pack
index_, which ranges from 0 to one less than the arity of the expansion. The
pack index is implicitly used as an index into the packs referred to by
each-names. This is easiest to see with statement pack expansions. For example,
if `a`, `x`, and `y` are packs with arity 3, then
`... each a += each x * each y;` is roughly equivalent to

```carbon
a[:0:] += x[:0:] * y[:0:];
a[:1:] += x[:1:] * y[:1:];
a[:2:] += x[:2:] * y[:2:];
```

Here we are using `[::]` as a hypothetical pack indexing operator for purposes

> of illustration; packs cannot actually be indexed in Carbon code.

Reviewer 4: Tiny nit: I find the `::`s run together a bit here, maybe:
```suggestion
Here we are using `[:N:]` as a hypothetical pack indexing operator for purposes
of illustration; packs cannot actually be indexed in Carbon code.
```

> 

Reviewer 4: Could we add a note that indexing is still open to exploration? I suspect folks coming from C++ might specifically be looking for that, and it seems good to not squash hope too much. Maybe:

```suggestion

> **Future work:** We're open to eventually adding indexing of variadics,
> but that remains future work and will need its own proposal.

```

Or something similar...

`⟬f32, Optional(each T), ⟪i32; ‖each y‖⟫⟭`. The `⟬⟭` delimiters make this a
_pack literal_ rather than a tuple literal. Notice one subtle difference: the
segments of a pack literal do not contain `...`. In effect, every segment of a
pack literal acts as a separate loop body. As with the tuple literal syntax, the
pack literal pseudo-syntax can also be used in patterns.

The _shape_ of a pack literal is a tuple of the arities of its segments, so the
shape of `⟬f32, Optional(each T), ⟪i32; ‖each y‖⟫⟭` is
`(1, ‖each T‖, ‖each y‖)`. Other expressions and patterns also have shapes. In
particular, the shape of an arity coercion `⟪E; A⟫` is `(A)`, the shape of
`each X` is `‖each X‖`, and the shape of an expression that does not contain
pack literals, shape coercions, or each-names is 1. The arity of an expression
is the sum of the elements of its shape. See the
[appendix](#typing-and-shaping-rules) for the full rules for determining the
shape of an expression.

A pack literal can be _expanded_, which moves its parent AST node inside the
pack literal, so long as the parent node is not `...`. For example,
`... Optional(⟬each X, Y⟭)` is equivalent to

> `... ⟬Optional(each X), Optional(Y)⟭`. Similarly, an arity coercion can be

Reviewer 4: I find "parent AST node" a strange concept to introduce here. This is a language design, and not really a compiler design, so it seems a bit circular to talk about AST nodes.

Maybe "moves its parent delimited syntactic construct inside the pack literal"? "deliminited syntactic construct" seems to be a reasonable description to cover the cases you care about here maybe?

Author: Hmm. I would read "delimited syntactic construct" to mean that there need to be delimiters, such as the parentheses in this example, but that's not the intent; the example could just as well be that `... 2 * ⟬each X, Y⟭` is equivalent to `... ⟬2 * each X, 2 * Y⟭`. "Syntactic construct" might work, but it seems awfully vague.

Would "syntax tree node" work better for you?

It's possible that any way of expressing this is going to seem a little strange, because the language rule I'm describing here is a little strange. Even the "formal" version of this rule in the appendix (starting around line 704) winds up being pretty handwavy, because it can't be expressed as a reduction rule; it's really a rule for generating reduction rules.

Author: I've now rephrased with a variant of the suggestion you made offline.

-   [Execution Semantics](#execution-semantics)
    -   [Expressions and statements](#expressions-and-statements)
    -   [Pattern matching](#pattern-matching)
-   [Typechecking](#typechecking)
    -   [Tuples, packs, segments, and shapes](#tuples-packs-segments-and-shapes)
    -   [Iterative typechecking of pack expansions](#iterative-typechecking-of-pack-expansions)
    -   [Typechecking patterns](#typechecking-patterns)
    -   [Typechecking pattern matches](#typechecking-pattern-matches)
-   [Appendix: Type system formalism](#appendix-type-system-formalism)
    -   [Explicit deduced arities](#explicit-deduced-arities)
    -   [Typing and shaping rules](#typing-and-shaping-rules)
    -   [Reduction rules](#reduction-rules)
    -   [Equivalence, equality, and convertibility](#equivalence-equality-and-convertibility)
    -   [Pattern match typechecking algorithm](#pattern-match-typechecking-algorithm)
        -   [Canonicalization algorithm](#canonicalization-algorithm)
-   [Alternatives considered](#alternatives-considered)
-   [References](#references)

<!-- tocstop -->

## Basics

### Overview

A "pack expansion" is a syntactic unit beginning with `...`, which is a kind of

> compile-time loop over sequences called "packs". Packs are initialized and

Reviewer 2: For what it's worth: I don't find it is helping my understanding to refer to this as a compile-time loop, or to describe the operation of pack expansion as iterative. The outcome would be the same if we processed pack elements in any order or in parallel, and the implied sequencing here is adding unnecessary complexity for me, even though I'd expect a typical implementation to be iterative.

Buuut... maybe this is the right way to describe this for other audiences than me :) And just so long as people think of this as "there is a loop in the compiler", not "there is a loop in the generated SemIR", I don't think it's giving the wrong impression. So, I'm not asking for a change here.

Author: Yeah, the tradeoffs of expressing this in procedural vs. functional terms are really complicated. Not only are some people more comfortable with one or the other, but different contexts work better with one or the other (e.g. pack expansion statements really lend themselves to a loop description, but the type system is much easier to express functionally).

I think the main reason I took the current approach is that the procedural description requires less up-front explanation, so it's a better fit for these introductory sections, but then I shift more to the functional description as we get more into the weeds, because it's much more suited to formal reasoning.

One thing I could potentially change here: I could say that pack expansions are _run-time_ loops, but they typically must be unrolled during type checking (because of heterogeneity and because of the tuple usage), which causes them to behave like compile-time loops as well. I think that's more consistent with how I talk about them in the rest of the doc (e.g. it would resolve the issue you flagged with `...and` and `...or` below). What do you think?

Reviewer 2: I think changing to talking about runtime loops would make me less happy, because of the cases you mentioned where additional handwaving is required, though perhaps that is more consistent with the exposition here. We can only really give a loop at runtime in the case where the pack is "sufficiently" homogeneous (in similar cases to when we could avoid monomorphization), and that seems to be enough of an edge case that it's probably misleading to center our description around it, even if we do want loop-like lowering for the sufficiently homogeneous cases.

referred to using "each-names", which are marked with the `each` keyword at the
point of declaration and the point of use.

The syntax and behavior of a pack expansion depends on its context, and in some
cases on a keyword following the `...`:

-   In a tuple literal expression (such as a function call argument list), `...`
    iteratively evaluates its operand expression, and treats the values as
    successive elements of the tuple.
-   `...and` and `...or` iteratively evaluate a boolean expression, combining
    the values using `and` and `or`, and ending the loop early if the underlying

>     operator short-circuits.

Reviewer 2: I previously asked about whether the loop terminates early at compile time, and you said it doesn't, so to my reading this is not saying what you intend -- rather, the compile-time loop does not end early, ever. (But the compile-time loop generates code that at runtime might short circuit.)

```suggestion
    the values using `and` and `or`. Normal short-circuiting behavior for the
    resulting `and` and `or` operators applies at runtime.
```

Reviewer 2: The other bullets in this section also talk about runtime things happening as part of the compile-time loop -- eg, in a tuple literal expression, the compile-time loop doesn't iteratively evaluate anything, because the evaluation typically happens at runtime -- but I think this instance is more likely to confuse readers than the others.

matched with the last M elements of the scrutinee. If the scrutinee does not
have at least N + M elements, the pattern does not match.

The remaining elements of the scrutinee are iteratively matched against
_operand_, in order. In each iteration, `$I` is equal to the index of the
scrutinee element being matched, minus N.

On the Nth iteration, a binding pattern binds the Nth element of the named pack
to the Nth scrutinee value.

## Typechecking

### Tuples, packs, segments, and shapes

In order to discuss the underlying type system for variadics, we will need to
introduce some pseudo-syntax to represent values and expressions that occur in
the type system, but cannot be expressed directly in user code. We will use
non-ASCII glyphs such as `⟪⟫‖⟬⟭` for that pseudo-syntax, to distinguish it from
valid Carbon syntax.

In the context of variadics, we will say that a tuple literal consists of a
comma-separated sequence of _segments_, and reserve the term "elements" for the
components of a tuple literal after pack expansion. For example, the expression
`(... each foo)` may evaluate to a tuple value with any number of elements, but

> the expression itself has exactly one segment.

Reviewer 2: It would have helped my understanding to say here that we're going to assign a generalized notion of a type to each segment, where the type of a segment describes (potentially symbolically) both the arity of the segment and the type of each expansion of the segment.

Author: How's this?

segment-wise. For example, suppose we are trying to find the type of `z` in this
code:

```carbon
fn F[... each T:! type]((... each x: Optional(each T)), (... each y: i32)) {
  let z: auto = (0 as f32, ... each x, ... each y);
}
```

We proceed by finding the type of each segment. The type of `0 as f32` is `f32`,
by the usual non-variadic typing rules. The type of `... each x` is
`... Optional(each T)`, because `Optional(each T)` is the declared type of
`each x`, and the type of a pack expansion is a pack expansion of the type of
its body.

The type of `... each y` is more complicated. Conceptually, it consists of some
number of repetitions of `i32`. We don't know exactly how many repetitions,
because it's implicitly specified by the caller: it's the arity of the second
argument tuple. Effectively, that arity acts as a hidden deduced parameter of
`F`.

So to represent this type, we need two new pseudo-syntaxes:

-   `‖each X‖` refers to the deduced arity of the pack expansion that contains
    the declaration of `each X`.
-   `⟪E; N⟫` evaluates to `N` repetitions of `E`. This is called a _arity
    coercion_, because it coerces the expression `E` to have arity `N`. `E` must

>     not contain any pack expansions, each-names, or pack literals (see below).

Reviewer 2: The Unicode symbols in use here are working badly for me both in github's code review flow and in its rendered markdown -- both the `‖` symbol and the `⟪` are rendering in two different ways in different contexts, and so it looks like there are four different sets of symbols in use here, not two. I don't know if this is a browser bug or not, but it's definitely making this harder to read and review.

Screenshot: note the much smaller `‖` in the paragraph starting "Combining the two", and that the definition of `⟪` appears to be a single chevron whereas the bottom line of the screenshot is clearly a double chevron.
<img width="590" alt="Screenshot 2024-09-29 at 13 22 37" src="https://github.com/user-attachments/assets/c392ff07-6eee-4c13-b0fe-0951489c16d3">

Author: Yep, I see that too. It also breaks monospacing.

This could be a browser bug, or a bug in the font itself, or a combination of the two. The deeper I dig into this, the more baffling it gets. The size of the characters depends not only on spatial context, but temporal context: In a text box, I can get them to change size by editing the surrounding text, in ways that persist even when I undo those edits. Fortunately those changes don't persist past saving the contents of the text box.

I haven't been able to find any options for fixing this directly, given that we don't control the font or the CSS. It looks like we could avoid the problem with other glyphs, but our options aren't great, because most of the math glyphs seem to have this problem to some extent. The best option I've found so far is:

`⸨f32, Optional(each T), ❰i32; ╎each y╎❱⸩`

`⟅⟆` also seem pretty well-behaved, and we might still be able to work with `‖` and `⟬` if we don't care about exact monospacing. Conversely, `╎` seems to behave itself in monospace rendering, but it sometimes acts hinky in the proportional-width editing text box, so I don't entirely trust it.

But I don't know if we want to let our notation be dictated by the vagaries of font rendering bugs; for all we know, whatever we choose could be broken the next time GitHub changes their CSS. That suggests we should either fall back on some ASCII notation (my previous approach was to use `{}<>||` and hope the theoretical ambiguities didn't come up too much in practice), or just live with these rendering glitches (and maybe file some bug reports). 

What do you suggest?

Reviewer 2: Let me try this:

> -   `‌‖‌each X‌‖‌` refers to the deduced arity of the pack expansion that contains
>     the declaration of `each X`.
> -   `‌⟪‌E; N‌⟫‌` evaluates to `N` repetitions of `E`. This is called a _arity
>     coercion_, because it coerces the expression `E` to have arity `N`. `E` must
>     not contain any pack expansions, each-names, or pack literals (see below).
> 
> Combining the two, the type of `... each y` is `... ‌⟪‌i32; ‌‖‌each y‌‖‌‌⟫‌`. Thus, the
> type of `z` is `(f32, ... Optional(each T), ... ‌⟪‌i32; ‌‖‌each y‌‖‌‌⟫‌)`.
> 
> Now, consider a modified version of that example:
> 
> ```carbon
> fn F[... each T:! type]((... each x: Optional(each T)), (... each y: i32)) {
>   let (.. each z: auto) = (0 as f32, ... each x, ... each y);
> }
> ```
> 
> `each z` is a pack, but it has the same elements as the tuple `z` in our earlier
> example, so we represent its type in the same way, as a sequence of segments:
> `‌⟬‌f32, Optional(each T), ‌⟪‌i32; ‌‖‌each y‌‖‌‌⟫‌‌⟭‌`. The `‌⟬‌‌⟭‌` delimiters make this a
> _pack literal_ rather than a tuple literal. Notice one subtle difference: the
> segments of a pack literal do not contain `...`. In effect, every segment of a
> pack literal acts as a separate loop body. As with the tuple literal syntax, the
> pack literal pseudo-syntax can also be used in patterns.
> 
> The _shape_ of a pack literal is a tuple of the arities of its segments, so the
> shape of `‌⟬‌f32, Optional(each T), ‌⟪‌i32; ‌‖‌each y‌‖‌‌⟫‌‌⟭‌` is
> `(1, ‌‖‌each T‌‖‌, ‌‖‌each y‌‖‌)`. Other expressions and patterns also have shapes. In
> particular, the shape of an arity coercion `‌⟪‌E; A‌⟫‌` is `(A)`, the shape of
> `each X` is `‌‖‌each X‌‖‌`, and the shape of an expression that does not contain

That seems to be working for me. What I did was to insert a U+200C (zero width non-joiner) on both sides of each of the fancy characters. Unfortunately that seems to give the worse rendering of the double-chevron. I wonder if using guillemets (U+AB, U+BB: « ») instead of the mathematical double-angle brackets would work better:

> `‌⟬‌f32, Optional(each T), ‌«i32; ‌‖‌each y‌‖‌‌»⟭‌`. The `‌⟬‌‌⟭‌` delimiters make this a

That seems OK. Maybe there's a better character to use here, though.

Author: I avoided guillemets and similar characters because their small vertical size interferes with visual scanning, but maybe that's OK for shape coercions since they're generally pretty small. In any event, it looks like most of the glyphs I rejected for that reason also screw up the rendering of `‖`. `‹›` also seem to work, but I like the visual consistency of all the variadic-related glyphs being doubled (ish).

I've converted the double-angles to guillemets. WDYT?

Combining the two, the type of `... each y` is `... ⟪i32; ‖each y‖⟫`. Thus, the
type of `z` is `(f32, ... Optional(each T), ... ⟪i32; ‖each y‖⟫)`.

Now, consider a modified version of that example:

```carbon
fn F[... each T:! type]((... each x: Optional(each T)), (... each y: i32)) {

>   let (.. each z: auto) = (0 as f32, ... each x, ... each y);

Reviewer 2: ```suggestion
  let (... each z: auto) = (0 as f32, ... each x, ... each y);
```

}
```

`each z` is a pack, but it has the same elements as the tuple `z` in our earlier
example, so we represent its type in the same way, as a sequence of segments:
`⟬f32, Optional(each T), ⟪i32; ‖each y‖⟫⟭`. The `⟬⟭` delimiters make this a
_pack literal_ rather than a tuple literal. Notice one subtle difference: the
segments of a pack literal do not contain `...`. In effect, every segment of a
pack literal acts as a separate loop body. As with the tuple literal syntax, the
pack literal pseudo-syntax can also be used in patterns.

The _shape_ of a pack literal is a tuple of the arities of its segments, so the
shape of `⟬f32, Optional(each T), ⟪i32; ‖each y‖⟫⟭` is
`(1, ‖each T‖, ‖each y‖)`. Other expressions and patterns also have shapes. In
particular, the shape of an arity coercion `⟪E; A⟫` is `(A)`, the shape of

> `each X` is `‖each X‖`, and the shape of an expression that does not contain

Reviewer 2: Why does `(A)` have parens but `‖each X‖` does not?

Author: I think that's just an oversight. Fixed (also added trailing commas, to ensure these are tuple literals).

> pack literals, shape coercions, or each-names is 1. The arity of an expression

Reviewer 2: ```suggestion
pack literals, shape coercions, or each-names is `1`. The arity of an expression
```
(Consistently using code font for shapes.)

Should this be `(1)`?

is the sum of the elements of its shape. See the
[appendix](#typing-and-shaping-rules) for the full rules for determining the
shape of an expression.

A pack literal can be _expanded_, which moves its parent AST node inside the

> pack literal, so long as the parent node is not `...`. For example,

Reviewer 2: If moving the parent node inside means also moving its other children inside, then this doesn't seem quite right for things like `... (⟬each X, Y⟭, ⟬each X, Y⟭)`, which should be expanded to `... ⟬(each X, each X), (Y, Y)⟭` not to `... ⟬(each X, ⟬each X, Y⟭), (Y, ⟬each X, Y⟭)⟭)`. But maybe you cover this later in the detailed rules.

Author: Yeah, the phrasing here is a simplification to try to get the core idea across. The formal (ish) rules below do cover this case (see "_Pack expanding_", around line 720).

    the type of its body.
-   The type of a tuple literal is a tuple literal consisting of the types of
    its segments.
-   If an expression or pattern `E` contains a pack literal or arity coercion
    that is not inside a pack expansion, the type of `E` is the type of the
    fully expanded form of `E`.

> **TODO:** address `...expand`, `...and` and `...or`.

### Reduction rules

Unless otherwise specified, all expressions in these rules must be free of side
effects. Note that every reduction rule is also an equivalence: the utterance
before the reduction is equivalent to the utterance after, so these rules can
sometimes be run in reverse (particularly during deduction).

Utterances that are reduced by these rules must be well-shaped (and the reduced
form will likewise be well-shaped), but need not be well-typed. This enables us
to apply these reductions while determining whether an utterance is well-typed,
as in the case of typing an expression or pattern that contains a pack literal
or arity coercion, above.

_Singular pack removal:_ if `E` is a pack segment, `⟬E⟭` reduces to `E`.

_Singular expansion removal:_ `...E` reduces to `E`, if `E` contains no pack

> literals, arity coercions, or each-names.

Reviewer 2: I found this rule surprising -- if we reach this point it seems like we've somehow lost the arity of `E`, and assuming that that means it should have an arity of 1 worries me a bit. But I think the overall logic here is that a segment either contains a construct with fixed arity, or implicitly has arity 1, so this is probably correct (but surprising).

Author: Yeah, it seems less surprising to state this in terms of the shape being `(1,)`, and I think it's equivalent, so I've made that change.

_Pack expansion splitting:_ If `E` is a segment and `S` is a sequence of
segments, `...⟬E, S⟭` reduces to `...E, ...⟬S⟭`.

_Pack expanding:_ If `F` is a function, `X` is an utterance that does not
contain pack literals or each-names, and `⟬P1, P2⟭` and `⟬Q1, Q2⟭` both have the
shape `(S1, S2)`, then `F(⟬P1, P2⟭, X, ⟬Q1, Q2⟭, ⟪Y; S1+S2⟫)` reduces to
`⟬F(P1, X, Q1, ⟪Y; S1⟫), F(P2, X, Q2, ⟪Y; S2⟫)⟭`. This rule generalizes in

> several dimensions:

Reviewer 2: It doesn't look to me like the case in pack expanding where `X` contains an arity coercion works correctly. Given `F(⟬P1, P2⟭, (⟪X; 2⟫), ⟬Q1, Q2⟭, (⟪Y; 1 + 1⟫)`, we seem to get a pack expansion to `⟬F(P1, (⟪X; 2⟫), Q1, (⟪Y; 1⟫), F(P2, (⟪X; 2⟫), Q2, (⟪Y; 1⟫)⟭` rather than to `⟬F(P1, (⟪X; 1⟫), Q1, (⟪Y; 1⟫), F(P2, (⟪X; 1⟫), Q2, (⟪Y; 1⟫)⟭`.

Is this intended to have a restriction that `X` does not contain an arity coercion? (And that there can be zero or more `Y` arity coercion arguments, not just exactly one?)

Author: > Is this intended to have a restriction that X does not contain an arity coercion?

Whoops, yes. Fixed.

> (And that there can be zero or more `Y` arity coercion arguments, not just exactly one?)

That was technically already covered, but I've made it more explicit.

## Table of contents

-   [Basics](#basics)
-   [Expression and statement semantics](#expression-and-statement-semantics)
-   [Typechecking expressions and statements](#typechecking-expressions-and-statements)
    -   [Generalized tuple types](#generalized-tuple-types)
    -   [Iterative typechecking](#iterative-typechecking)
    -   [In-place typechecking](#in-place-typechecking)
-   [Pattern semantics](#pattern-semantics)
-   [Typechecking pattern matching](#typechecking-pattern-matching)
    -   [Identifying potential matchings](#identifying-potential-matchings)
    -   [The type-checking algorithm](#the-type-checking-algorithm)
-   [Alternatives considered](#alternatives-considered)
-   [References](#references)

<!-- tocstop -->

> ## Basics

Reviewer: I think there is a bigger change that would be good to make to this section to improve the exposition. However, it doesn't seem necessary to block this PR on. I'm going to restrict to more tactical comments for now, and we can tackle bigger changes in a follow-up.

keyword at the point of declaration and the point of use.

The syntax and behavior of a pack expansion depends on its context, and in some
cases by a keyword following the `...`:

-   In a tuple literal expression, `...` iteratively evaluates its operand
    expression, and treats the values as successive elements of the tuple.
-   `... and` and `... or` iteratively evaluate a boolean expression, combining
    the values using `and` and `or`, and ending the loop early if the underlying
    operator short-circuits.
-   In a statement context, `...` iteratively executes a statement.
-   In a tuple literal pattern, `...` iteratively matches the elements of the
    scrutinee tuple. In conjunction with variadic bindings, this enables
    functions to take an arbitrary number of arguments.

This example illustrates many of the key concepts:

```carbon
// Takes an arbitrary number of vectors with arbitrary element types, and
// returns a vector of tuples where the i'th element of the vector is
// a tuple of the i'th elements of the input vectors.
fn Zip[... each ElementType:! type]
      (... each vector: Vector(each ElementType))
      -> Vector((... each ElementType)) {

>   ... var each iter: auto = each vector.Begin();

Reviewer: A bit weird that it is `... each` in the function signature, but `... var each` in the function body. Seems like `var` introduces a pattern context and so `var ... each` would be the way of declaring a pack variable. I guess that we need the `...` before the `var` since we are also expanding the initializer (and could be expanding the variable type)? I guess that means `var ... each foo: i32 = something;` would never be legal?

Might be worth a comment pointing out the difference.

Author: > A bit weird that it is `... each` in the function signature, but `... var each` in the function body. Seems like `var` introduces a pattern context and so `var ... each` would be the way of declaring a pack variable. I guess that we need the `...` before the `var` since we are also expanding the initializer (and could be expanding the variable type)?

Right, we don't want to repeat just the pattern, we want to repeat the entire statement.

> I guess that means `var ... each foo: i32 = something;` would never be legal?

Yes, but for a different reason: it doesn't correspond to any of the valid pack expansion syntaxes listed below. I think something like this would be legal, though:

```
var (... each foo: i32) = (... each vector.Begin());
```

But that seems like just a more complicated version of what I have here.

> Might be worth a comment pointing out the difference.

I think it would get unwieldy to try to comment this example to point out all the concepts it's illustrating, because this example is intentionally dense. The example is unavoidably going to be hard to follow on a first read, because I haven't explained how the features in the example work yet. I could put the example afterward, but I think the prose explanation would seem really abstract and hard to follow without an example to refer to.

In the longer run I think the way out of the dilemma is your suggestion of building up the exposition more gradually from a series of simpler examples, but we're deferring that for now.

same precedence as `,`, `and`, and `or`, respectively. They are single tokens,
and may not contain internal whitespace. `...{` is an opening delimiter that
forms a balanced pair with `}`, and can contain any number of statements. An AST

> rooted at any of these operations is called a _pack expansion_. The operand of

Reviewer: Seems weird that you define "pack expansion" but not "pack"

Author: Agreed. I've been somewhat cautious about changing terminology because it could confuse the discussion, but this might be a place where it's worth doing. Do you think "expansion" would work?

Note that the "pack" concept still exists, but it's been de-emphasized so much that it doesn't make sense to introduce it until much later.

Reviewer: I think it is probably better to leave "pack expansion" as it is, and plan on eventually defining "pack." That could be done in this PR, but it isn't necessary to do in this PR if you don't think it would help (though I suspect it would be helpful to at least decide on what it means and give some examples of packs so we can be all on the same page).

`...,`, `...and`, or `...or`, or the contents of the block delimited by
`...{ }`, is called the _expansion body_.

A pack expansion must contain one or more _expansion arguments_, which are

> marked by `expand` or `each`. `expand` is a prefix unary expression operator

Reviewer: This seems incomplete since `each` is also used in `each ElementType:! type` outside of a pack expansion in the example. Maybe have a note here that `each` has another meaning in a binding pattern?

Author: OK, how's this?

expansion must be expressions.

> **Open question:** Is it possible to drop that requirement, and support code

> > like `let each x: ElementType = *expand iters;` within a `...{` expansion?

Reviewer: I'm not sure what meaning this example code would supposedly have.

Author: Yeah, this seems nonsensical. Deleted.

I think I was trying to address the issue that's now covered by p2240.md#support-variadic-variable-declaration-statements, but this isn't the way to do it.

An expression of the form "`...and` _operand_" is evaluated as follows: a
notional `bool` variable `$R` is initialized to `true`, and then "`$R = $R and`
_operand_" is executed up to N times, with `$I` ranging from 0 to N - 1. If at
any point `$R` becomes false, this iteration is terminated early. The final
value of `$R` is the value of the expression.

An expression of the form "`...or` _operand_" is evaluated the same way, but
with `or` in place of `and`, and `true` and `false` transposed.

An expression of the form "`...,` _operand_" evaluates to a sequence of N
values, where the k'th value is the value of _operand_ where `$I` is equal to
k - 1.

An expression of the form "`expand` _argument_" evaluates to "_argument_
`[$I]`".

An expression of the form "`each` _identifier_", where _identifier_ names a
variadic binding, evaluates to the `$I`th value that it was bound to (indexed
from zero).

## Typechecking expressions and statements

### Generalized tuple types

The `...,` operator lets us form tuples out of sequences whose size is not known
during typechecking. For example, in this code:

```carbon
fn F[each T:! type](x: (..., each i32), ..., each y: Optional(each T)) {

>   let z: auto = (..., expand x, 0 as f32, each y);

Reviewer: This seems strange to me. It seems like there should be two pack expansions, if for no other reason than the number of elements of `x` is different from how many different `y` there are. I would expect this to be written:

```suggestion
  let z: auto = (..., expand x, 0 as f32, ..., each y);
```

Author: Whoops, yes.

}
```

The type of `z` is a tuple whose elements are an indeterminate number of
repetitions of `i32`, followed by `f32`, followed by a different indeterminate
number of types that all have the form `Optional(T)` for some type `T`. We can't
represent this as an explicit list of element types until those indeterminate
numbers are known, so we need a more general representation for tuple types.

In this model, a tuple type consists of a sequence of _pack components_, and a
pack component consists of a type called the _representative_, and an arity,
both of which may be symbolic. There is a special symbolic variable called the
_pack index_, which can only be used in the representative of a pack component,
and only as the subscript of an indexing expression on a tuple. The pack index
is used solely as a placeholder, and never has a value. There are also symbolic
variables representing the arity of every variadic binding. For purposes of
illustration, the notation `<V, N>` represents a pack component with

> representative V and arity N, `$I` represents the pack index, and given a

Reviewer: ```suggestion
representative `V` and arity `N`, `$I` represents the pack index, and given a
```

variadic binding `B`, `|B|` represents the arity of `B`, and `B/$I` represents
the `$I`th value bound by `B`.

So, continuing the earlier example, the type of `z` is represented symbolically

> as `(<i32, Sizeof(x)>, <f32, 1>, <Optional(T/$I), |T|>)`.

Reviewer: ```suggestion
as `(<i32, SizeOf(x)>, <f32, 1>, <Optional(T/$I), |T|>)`.
```

A pack component is _variadic_ if its arity is unknown, and _singular_ if its
arity is known to be 1 and its representative does not refer to the pack index.
In contexts where all pack components are known to be singular, we will
sometimes refer to them as "elements". Pack values are always assumed to be
normalized, meaning that every component is either singular or variadic. This is
always possible because if a component's arity is known to be some fixed value N
other than 1, we can replace it with N singular components. The _shape_ of a
tuple type is the sequence of arities of its components, so the shape of the

> type of `z` is `(Sizeof(x), 1, |T|)`.

Reviewer: ```suggestion
type of `z` is `(SizeOf(x), 1, |T|)`.
```

after another. As a result, a pack type is never the type of any run-time value.

## Pattern semantics

`,...` expansions can also appear in patterns. The semantics are chosen to
follow the general principle that pattern matching is the inverse of expression
evaluation, so for example if the pattern `(..., expand x: auto)` matches some
scrutinee value `s`, the expression `(..., expand x)` should be equal to `s`.
These semantics are implemented at monomorphization time, so all types are known
constants, and all tuple elements are singular.

A tuple pattern can contain no more than one subpattern of the form "`...,`
_operand_". When such a subpattern is present, the N elements of the pattern
before the `...,` expansion are matched with the first N elements of the
scrutinee, and the M elements of the pattern after the `...,` expansion are
matched with the last M elements of the scrutinee. If the scrutinee does not
have at least N + M elements, the pattern does not match.

The remaining elements of the scrutinee are iteratively matched against
_operand_, in order. In each iteration, `$I` is equal to the index of the

> scrutinee element being matched.

Reviewer: Minus N?

A variadic binding pattern binds the name to each of the scrutinee values, in
order.

> A pattern of the form `expand` _subpattern_ matches the `$I`th element of

Reviewer: Example?

Author: Do you just want a syntactic example of this kind of pattern, or do you want a discussion of how these semantic rules apply to the example (and if so, which of the three cases would be most helpful to illustrate)? Does the example need to be motivating, or can it be contrived to illustrate the rules?

which we will also refer to as the arity of the expansion. If an expansion
contains no each-names, it must be a pattern, or an expression in the type
position of a binding pattern, and its arity is deduced from the scrutinee.

A pack expansion or `...expand` expression cannot contain another pack expansion
or `...expand` expression.

An each-name cannot be used in the same pack expansion that declares it. In most
if not all cases, an each-name that violates this rule can be changed to an
ordinary name, because each-names are only necessary when you need to transfer a
pack from one pack expansion to another.

A pack expansion expression or statement can be thought of as a kind of loop
that executes at compile time (specifically, monomorphization time), where the
expansion body is implicitly parameterized by an integer value called the _pack
index_, which ranges from 0 to one less than the arity of the expansion. The
pack index is implicitly used as an index into the packs referred to by
each-names. This is easiest to see with statement pack expansions. For example,
if `a`, `x`, and `y` are packs with arity 3, then
`... each a += each x * each y;` is roughly equivalent to

```carbon

> for (let i:! i32 in (0, 1, 2)) {

Reviewer: Expressing this as a loop seems a bit misleading since there is no new scope introduced. It seems like this is more equivalent to
```
a[:0:] += x[:0:] * y[:0:];
a[:1:] += x[:1:] * y[:1:];
a[:2:] += x[:2:] * y[:2:];
```
This also addresses the type checking concern brought up in the next paragraph (though perhaps introduces new ones).

Author: OK, done.

What new concerns do you have in mind?

A pack expansion pattern "`...` _subpattern_" appears as part of a tuple pattern
(or an implicit parameter list), and matches a sequence of tuple elements if
each element matches _subpattern_. For example, in the signature of `Zip` shown
earlier, the parameter list consists of a single pack expansion pattern

> `... each vector: Vector(each ElementType)`.

Reviewer: ... and so each argument type will be matched against `each vector: Vector(each ElementType)`.

Author: I've tweaked the phrasing a bit. How's this?

Since _subpattern_ will be matched against multiple scrutinees (or none) in a
single pattern-matching operation, a binding patterns within a pack expansion

> pattern must declare an each-name, and the Nth iteration of the pack expansion

Reviewer: ```suggestion
pattern must declare an each-name (`each vector` in the `Zip` example), and the Nth iteration of the pack expansion
```

will initialize the Nth element of the named pack from the Nth scrutinee. The

> binding pattern's type expression may contain an each-name, but if so, it must

Reviewer: ```suggestion
binding pattern's type expression may contain an each-name (as in `each ElementType` in the `Zip` example), but if so, it must
```

`(F(x[:0:], y[:0:]), F(x[:1:], y[:1:]), F(x[:2:], y[:2:]))`. This can't be
expressed as a loop in Carbon code, but it is still fundamentally iterative.

> A pack expansion pattern "`...` _subpattern_" appears as part of a tuple pattern

Reviewer: Can you show this with an example?

Author: OK, how's this?

-   For a `...` tuple element expression or pattern, the segments of the
    operand's type pack become segments of the type of the enclosing tuple.

> > **TODO:** Discuss typechecking `...expand`.

Reviewer: My understanding is that you are not planning to execute any of these TODOs in this proposal, but in a follow-up PR. Is that correct? (If you are planning to do them, it might make sense to use a different marker, like "FIXME".)

Author: That's right.

### Typechecking patterns

A _full pattern_ consists of an optional deduced parameter list, a pattern, and
an optional return type expression.

A pack expansion pattern has _fixed arity_ if it contains at least one usage of
an each-name that is not a parameter of the enclosing full pattern. Otherwise it
has _deduced arity_. A tuple pattern can have at most one segment with deduced

> arity.

Reviewer: Can you show examples of each so it is clear how this difference arises?

```
class FixedArityExample(... each T: type) {
  // Arity is determined by the arguments to the class,
  // not the arguments to the call to this function.
  fn F(...each x: each T);
}

Author: How's this?

must match at least one element.)

When the pattern is heterogeneous, the merging process may be more complex. For
example:

```carbon
fn ZipAtLeastOne[First:! type, ... each Next:! type]
    (first: Vector(First), ... each next: Vector(each Next))
    -> Vector((First, ... each Next));
```

During typechecking, we transform that function signature to the following form:

```carbon
fn ZipAtLeastOne[... ⟬First, each Next⟭:! type]
    (... each args: Vector(⟬First, each Next⟭))
    -> Vector((...⟬First, each Next⟭));
```

In this rewritten form, we treat the pack of deduced parameter names
`⟬First, each Next⟭` as though it were the each-name of a single deduced
parameter, called a _synthetic deduced parameter_. This enables us to model this
function as only having one parameter. The fact that it's composed of an
ordinary name and an each-name has no further effect, except to record the fact
that this pack must have at least one element (and consequently `each args` must
too). Note in particular that `Vector(⟬First, each Next⟭)` is considered to be
fully expanded, and this it has a single scalar component, itself.

> A synthetic deduced parameter must satisfy the following requirements:

Reviewer: Am I missing a requirement that all uses are in the same order? So we would reject
```
fn Reorders[First:! type, ... each Next:! type]
    (first: Vector(First), ... each next: Vector(each Next))
    -> Vector((.. each Next, First));
```

Author: That's a requirement on the rewritten signature as a whole, so it's in the second bulleted list. I've added an example illustrating that.

-   Its segments are all the names of deduced parameters of the full pattern.
-   No name occurs more than once.
-   One of the names must be an each-name.

> -   All the names must have the same declared type.

Reviewer: For each requirement, can you give an example that fails the requirement?

Author: Done.

The rewritten full pattern must satisfy the following requirements:

-   Names that are part of a synthetic deduced parameter are not used outside a
    synthetic deduced parameter, or in synthetic deduced parameters that are not
    equal.
-   A pack expansion containing a synthetic deduced parameter doesn't contain

>     any pack literals that aren't synthetic deduced parameters.

Reviewer: Examples showing how these can fail would be good too.

Author: Done.

See the [appendix](#deduction-algorithm) for a more formal discussion of the
rewriting process.

### Typechecking pattern matches

To typecheck a pattern match between a tuple pattern and a tuple scrutinee, we
try to split and merge the segments of the scrutinee type so that it has the
same number of segments as the pattern type, and corresponding segments have the
same arity. For example, consider this call to `ZipAtLeastOne` (as defined in
the previous section):

```carbon
fn F[... each T:! I, U:! J](... each t: Vector(each T), u: Vector(U)) {
  ZipAtLeastOne(... each t, u);
}
```

The pattern type is `(... Vector(⟬First, each Next⟭))`, so we need to rewrite
the scrutinee type `(... Vector(each T), Vector(U))` to have a single segment
with an arity that matches `‖each Next‖+1`. We can do that by merging the
scrutinee segments in much the same way that we originally did with the pattern
segments, to obtain `(... Vector(⟬each T, U⟭))`. This has a single segment with
arity `‖each T‖+1`, which can match `‖each Next‖+1` because the deduced arity
`‖each Next‖` behaves as a deduced parameter of the pattern, so they match by

> deducing `‖each Next‖ == ‖each T`.

Reviewer: ```suggestion
`‖each Next‖` behaves as a deduced parameter of the pattern, so they match by
deducing `‖each Next‖ == ‖each T‖`.
```

> pattern context, so name-packs-as-names have different semantics.

## Appendix: Type system formalism

### Typing and shaping rules

The _body arity_ of a pack expansion pattern is determined as follows:

-   If it contains a usage of at least one each-name that is not a parameter of
    the pattern, the body arity is the shape of that each-name.
-   Otherwise, the body arity is a hidden deduced parameter of the pattern
    called the _deduced arity binding_.

The shapes of expressions and statements within a pack expansion are determined
as follows:

-   The shape of an arity coercion is the value of the expression after the `;`.
-   The shape of a pack literal is the concatenation of the arities of its
    segments.
-   The shape of a binding pattern that declares an each-name is the body arity
    of the enclosing pack expansion.
-   The shape of an each-name expression is the body arity of the pack expansion
    containing the declaration of the each-name.
-   Any other AST node is _well shaped_ if there is some shape `S` such that all
    of the node's children have shape either 1 or `S`. When that condition

>     holds, the shape of the node is `S` (or 1 if all children have shape 1).

Reviewer: I don't think we use parentheticals for load-bearing conditions.
```suggestion
    holds, the shape of the node is 1 if all children have shape 1, otherwise `S`.
```

    declared it.
-   The type of an arity coercion `⟪E; S⟫` is `⟪T; S⟫`, where `T` is the type of
    `E`.
-   The type of a pack literal is a pack literal consisting of the concatenated
    types of its segments. This concatenation flattens any nested pack literals
    (for example `⟬A, ⟬B, C⟭⟭` becomes `⟬A, B, C⟭`)
-   The type of a pack expansion expression or pattern is `...B`, where `B` is
    the type of its body.
-   The type of a tuple literal is a tuple literal consisting of the types of
    its segments.
-   If an expression `E` contains a pack literal or arity coercion that is not
    inside a pack expansion, the type of `E` is the type of the fully expanded
    form of `E`.

> **TODO:** address `...expand`, `...and` and `...or`.

### Reduction rules

Unless otherwise specified, all expressions in these rules must be free of side
effects. Note that every reduction rule is also an equivalence: the expression
before the reduction is equivalent to the expression after, so these rules can
sometimes be run in reverse (particularly during deduction).

Expressions that are reduced by these rules must be well-shaped (and the reduced

> form will likewise be well-shaped), but need not be well-typed.

Reviewer: Could you expand on the "not well-typed" situation?

Author: Sure, how's this?

-   `F()` can be any expression syntax other than `...` or pack literal
    formation, not just a function call.

_Coercion removal:_ `⟪E; 1⟫` reduces to `E`.

_Tuple indexing:_ Let `I` be an integer template constant, let `X` be a tuple
segment, and let `Ys` be a sequence of tuple segments.

-   If the arity `A` of `X` is less than `I+1`, then `(X, Ys).(I)` reduces to
    `(Ys).(I-A)`.
-   Otherwise:
    -   If `X` is not a pack expansion, then `(X, Ys).(I)` reduces to `X`.
    -   If `X` is of the form `...⟬⟪E; S⟫⟭`, then `(X, Ys).(I)` reduces to `E`.

### Other equivalences

Unless otherwise specified, all expressions in these rules must be free of side
effects.

_Coercion merging:_

-   `⟪E; M⟫, ⟪E; N⟫` is equivalent to `⟪E; M, N⟫`.
-   `⟪E; N⟫, E` is equivalent to `⟪E; N, 1⟫`.
-   `E, ⟪E; N⟫` is equivalent to `⟪E; 1, N⟫`.

_Coercion shape commutativty:_ `⟪E; S1⟫` is equivalent to `⟪E; S2⟫` if `S1` is a
permutation of `S2`.

### Convertibility and parameter deduction

Type convertibility is governed by a set of deduction rules for the "convertible
to" and "deducible from" relations. For example:

-   `T` is convertible to `U` if `T` implements `ImplicitAs(U)`.

> -   `T` is convertible to `U` if `U` is deducible from `T`.

Reviewer: When does this happen with `T != U`?

Author: I'm not sure I understand the question, but I've revised this section to try to make it clearer (and especially make the role of equality clearer). Maybe that will answer your question? If not, can you elaborate?

Reviewer: I mean "`U` is deducible from `T`" is usually accomplished by creating a binding map that makes `U` and `T` equal. My question is how could `U` be deduced from `T` and end up in a situation that `T` is convertible to `U` but not equal.

Author: OK, then the answer is that can't happen.

-   For any `X`, `X` is deducible from `X`.
-   Let `T` and `U` be tuple segments, and let `Ts` and `Us` be sequences of
    tuple segments. `(T, Ts)` is convertible to `(U, Us)` if `T` is convertible
    to `U` and `(Ts)` is convertible to `(Us)`.
-   Let `P` be a parameterized type. `P(A, B)` is deducible from `P(C, D)` if
    `A` is deducible from `C` and `B` is deducible from `D`.

Formally, these rules all implicitly propagate a _binding map_. For example, the
last rule above can be stated more precisely as:

-   Let `P` be a parameterized type. If `A` is deducible from `C` given a
    binding map `M1`, and `B` is deducible from `D` given a binding map `M2`,

>     then `P(A, B)` is deducible from `P(C, D)` given a binding map `M1` ∪ `M2`.

Reviewer: This seems like it could be true, but is not obvious that there is no way for `M1` and `M2` to have conflicting bindings. I couldn't find a counter example, but it seems to rely on something non-obvious deriving from the fact that `A, B` and `C, D` both satisfy the parameter constraints from `P`.

Author: As discussed at yesterday's open discussion, we can deduce conflicting bindings in situations like trying to deduce `P(T, T)` from `P(i32, i64)`.

A binding map is a set of pairs where the first element of the pair is a deduced
parameter of the pattern, and the second element is the deduced value of that
parameter. We model a parameter as a sequence of names, in order to accommodate
synthetic deduced parameters. A binding map must be _consistent_, which means
that if `N1` and `N2` are the names of two different name/value pairs in the

> map, `N1` cannot be a (non-strict) subsequence of `N2`.

Reviewer: Couldn't it still be consistent if the values have the same subsequence relationship as the names?

Author: I think that could work, but I don't think this will come up in practice, so I'd rather leave it out for simplicity.

In almost all cases, the binding map in the conclusion is the union of the
binding maps in the premises, and in those cases we will usually leave the
binding maps implicit (as in the original formulation of the above rule). One
major exception is the following rule:

_Binding introduction:_ Let `X` be a deduced parameter with type `T`. If the
type of `E` is convertible to `T` given binding map `M`, then `X` is deducible

> from `E` given a binding map `M` ∪ "`X` is bound to `E`".

Reviewer: It is a bit unclear what the role of `E` is in this rule. Maybe an example would help?

Author: It's hard to give an example of this rule in isolation, but I've added a worked example below that includes an application of this rule.

previous section. If that succeeds, we apply the resulting binding map to the
function return type to obtain the type of the call expression.

> **TODO:** Specify the algorithm more precisely. In particular, discuss how to
> rewrite `A` as needed to make the shapes line up, but don't rewrite `F` after
> canonicalization.

Typechecking for pattern match operations other than function calls is defined
in terms of typechecking a function call: We check a scrutinee type `S` against
a pattern `P` by checking `(S,)` against `(P,)->()`.

> **Future work:** Extend this approach to support merging the argument list as
> well as the parameter list.

#### Canonicalization algorithm

The canonical form can be found by starting with a normal form, and
incrementally merging an adjacent singular parameter type into the variadic
parameter type.

For example, consider the following function:

```carbon
fn F[First:! type, Second:! type, ... each Next:! type]
    (first: Vector(First), second: Vector(Second),
     ... each next: Vector(each Next)) -> (First, Second, ... each Next);
```

The function type derived from that signature is:

```carbon

> (Vector(First), Vector(Second), ... Vector(each Next))

Reviewer: Doesn't the function type also have deduced parameters?

Author: I've been thinking of the deduced parameter list as part of the function type's context, rather than part of the type itself. I haven't been able to figure out how to model deduced parameters in a function type, at least in this kind of formalism. In fact, the main reason I resorted to "synthetic deduced parameters" was to avoid the need to model rewriting the deduced parameter list.

If you have any suggestions for how to do this better, I'd love to hear them.

Reviewer: There are two kinds of functions: primitive function calls and families of function calls. A primitive function call has concrete parameter arity and types and no deduced parameters. Since you are concerned with variadic functions, primitive function calls are not enough, you need to be concerned about the families of function calls you get by having deduced parameters.

I grappled with this issue in [the "Compile-time function parameters" section of "Carbon: function overloading"](https://docs.google.com/document/d/1rTsDJ_ZNENH0mhi5E64MGjKJ29OYiI_dzBRz2GzYySU/edit?resourcekey=0-z6_xQ9xuJ5CZ6qJqiuSxuw&tab=t.0#heading=h.kx0d18wr7ba0). Variadics are trying to do the same thing: describe a family of implementations of the `Call` operator for a particular type, so:

```
fn F[... each T:! type](... each x: each T);
```

becomes:

```
class __typeof_F {}
let F:! __typeof_F = {}; 
final impl forall [... each T:! type] __typeof_F as Call(... each x: each T);
```

and then a particular call to `F` does "overload resolution" by looking to see if a matching implementation of `Call` exists for `__typeof_F`:

```
let a: i32 = 1;
let b: bool = true;
F(a, b);
```

is translated to:

```
let a: i32 = 1;
let b: bool = true;
F.((__typeof_F as Call(i32, bool)).Op)(a, b);
```

which finds the implementation of `Call` for `__typeof_F` with `each T` equal to `i32`, `bool`.

So the type of `F` is straightforward (`__typeof_F`), but not interesting (it is an empty class). It encapsulates all the overloads, but in a trivial way, since all values of that type are the same. What you are probably talking about, when you think of "function type", is some requirement that this type satisfies that says it is callable with some argument types, as in:

```
fn HasFnArg[T:! Call(i32, bool)](f: T) {
  let a: i32 = 1;
  let b: bool = true;
  f(a, b);
}

HasFnArg(F);
```

However, there is no way to say `T` implements a family of implementations of `Call` -- which might be written like `fn HasGenericFnArg[T:! forall [U:! type] Call(U, U) where .Result = Vector(U)](f: T)` -- and that is not expected to change soon. One issue is that it is hard in general to prove that a particular type would satisfy this condition -- for example, there could be a specialization somewhere that implements `Call` for that type with a different return type (and it wouldn't even necessarily be visible at a call site to `HasGenericFnArg`). Functions are the rare case where we might be able to do it, since behind the scenes they define a `final` implementation of `Call` (which corresponds to only supporting "closed" overloading by default). The other big issue is even with a `final` implementation, it would still be quite difficult to validate that a particular `forall` condition is satisfied, which violates the principle that the compiler shouldn't be asked to solve hard problems.

So, I would avoid talking about "function types" as a thing that captures how a function can be called. Instead, just talk about the signature of a particular function declaration.

Author: > So, I would avoid talking about "function types" as a thing that captures how a function can be called. Instead, just talk about the signature of a particular function declaration.

OK, done. That turned out to be less messy than I feared, and it even lets me get rid of "synthetic deduced parameters".

A pack expansion pattern has _fixed arity_ if it contains at least one usage of
an each-name that is not a parameter of the enclosing full pattern. Otherwise it
has _deduced arity_. A tuple pattern can have at most one segment with deduced
arity. For example:

```carbon

> class C(... each T:! type)

Reviewer: ```suggestion
class C(... each T:! type) {
```

-   A pack expansion containing a synthetic deduced parameter doesn't contain
    any pack literals that aren't synthetic deduced parameters.

For example `⟬X, each Y⟭` is not a valid synthetic deduced parameter in either

> of the following function signatures:

Reviewer: Please say why not, include which rule is violated.

```carbon

> fn F[X:! type, ... each Y:! type](... each args: ⟬X, each Y⟭) -> ⟬each Y, X⟭;

Reviewer: That return type doesn't look valid. Is it supposed to be a tuple type?

### Convertibility and parameter deduction

Type convertibility is governed by a set of inference rules for the "convertible
to" relation. For example:

-   _Equality implies convertibility:_ `T` is convertible to `U` if `T` equals
    `U`.
-   _ImplicitAs rule:_ `T` is convertible to `U` if `T` implements
    `ImplicitAs(U)`.
-   _Tuple convertibility:_ Let `T` and `U` be tuple segments, and let `Ts` and
    `Us` be sequences of tuple segments. `(T, Ts)` is convertible to `(U, Us)`

>     if `T` is convertible to `U` and `(Ts)` is convertible to `(Us)`.

Reviewer: Is this just trying to express that each segment is convertible using induction?

Author: Yes, exactly.

Reviewer: How about just say:
```suggestion
-   _Tuple convertibility:_ Let `Ts` and `Us` be sequences of tuple segments. `(Ts)` is convertible to `(Us)` if the segments of `Ts` are convertible to the corresponding segments of `(Us)`.
```

Author: Done, with an edit to state that there must be equal numbers of segments.

    equals `C` given a binding map `M1`, and `B` equals `D` given a binding map
    `M2`, and `M1` ∪ `M2` is consistent, then `P(A, B)` equals `P(C, D)` given a
    binding map `M1` ∪ `M2`.

Almost all of the deduction rules can be generalized in the same way: the
binding map in the conclusion is the union of the binding maps in the premises,
and there's an additional premise that this union is consistent. If there are no
"equals" or "convertible to" relations in the premises, the binding map in the
conclusion is empty. When that pattern applies, we will usually leave the
binding maps implicit (as in the original formulation of the above rule). One
major exception is the following rule:

_Binding introduction:_ Let `X` be a deduced parameter with type `T`. If the
type of `E` is convertible to `T` given binding map `M`, and `M` ∪ {(`X`, `E`)}
is consistent, then `X` equals `E` given a binding map `M` ∪ {(`X`, `E`)}.

The deduction rules that are specific to variadic types are as follows:

-   _Expansion convertibility:_ `...T` is convertible to `...U` if the shape of
    `U` equals the shape of `T`, and each scalar component of `T` is convertible

>     to every scalar component of `U`.

Reviewer: This is saying that you don't match up the components and see that they are convertible, you check that the full outer product of scalar components is convertible? If so, could you add some words to make that explicit?

Author: That's correct. Is this clearer?

The deduction rules that are specific to variadic types are as follows:

-   _Expansion convertibility:_ `...T` is convertible to `...U` if the shape of
    `U` equals the shape of `T`, and all scalar components of `T` are

>     convertible to all scalar components of `U`.

Reviewer: ```suggestion
    each convertible to all scalar components of `U`.
```

Author: "**all** scalar components of `T` are **each** convertible..." feels awkwardly redundant to me. How about "the scalar components of `T` are each convertible..."?

Reviewer: Sure

    (first: Vector(First), ... ⟬second, each next⟭: Vector(⟬Second, each Next⟭))
    -> (First, ... ⟬Second, each Next⟭);
// Pack renaming
fn F[__N:! Arity, First:! type, ... each __A:! ⟪type; __N+1⟫]
    (first: Vector(First), ... each __a: Vector(each __A))
    -> (First, ... each __A);
```

This brings us back to a normal form, while reducing the number of tuple
segments. We can now repeat that process to merge the remaining parameter type:

```carbon
fn F[__N:! Arity, First:! type, ... ⟬each __A:! ⟪type; __N+1⟫⟭]
    (first: Vector(First), ... each __a: Vector(⟬each __A⟭))
    -> (First, ... ⟬each __A⟭);
// Pack expanding
fn F[__N:! Arity, First:! type, ... ⟬each __A:! ⟪type; __N+1⟫⟭]
    (first: Vector(First), ... each __a: ⟬Vector(each __A)⟭)
    -> (First, ... ⟬each __A⟭);
// Pack expanding
fn F[__N:! Arity, First:! type, ... ⟬each __A:! ⟪type; __N+1⟫⟭]
    (first: Vector(First), ... ⟬each __a: Vector(each __A)⟭)
    -> (First, ... ⟬each __A⟭);
// Pack expansion splitting (in reverse)
fn F[__N:! Arity, ... ⟬First:! type, each __A:! ⟪type; __N+1⟫⟭]
    (... ⟬first: Vector(First), each __a: Vector(each __A)⟭)
    -> (... ⟬First, each __A⟭);
// Pack expanding (in reverse)

> fn F[__N:! Arity, ... ⟬First, each __A⟭:! ⟪type; __N+2⟫⟭]

Reviewer: Mismatch in doubled parens
```suggestion
fn F[__N:! Arity, ... ⟬First, each __A⟭:! ⟪type; __N+2⟫]
```

    (... ⟬first, each __a⟭: ⟬Vector(First), Vector(each __A)⟭)
    -> (... ⟬First, each __A⟭);
// Pack expanding (in reverse)

> fn F[__N:! Arity, ... ⟬First, each __A⟭:! ⟪type; __N+2⟫⟭]

Reviewer: ```suggestion
fn F[__N:! Arity, ... ⟬First, each __A⟭:! ⟪type; __N+2⟫]
```

    (... ⟬first, each __a⟭: Vector(⟬First, each __A⟭))
    -> (... ⟬First, each __A⟭);
// Pack renaming

> fn F[__N:! Arity, ... __B:! ⟪type; __N+2⟫⟭]

Reviewer: ```suggestion
fn F[__N:! Arity, ... __B:! ⟪type; __N+2⟫]
```
